// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/heatMap/HeatmapDescriptor.js
import { __extends } from "tslib";
import { BucketNames, isDrillIntersectionAttributeItem } from "@gooddata/sdk-ui";
import { PluggableHeatmap } from "./PluggableHeatmap";
import { BigChartDescriptor } from "../BigChartDescriptor";
import { drillDownFromAttributeLocalId } from "../../../utils/ImplicitDrillDownHelper";
import { addIntersectionFiltersToInsight, modifyBucketsAttributesForDrillDown } from "../drillDownUtil";
import { executionConfigInsightConversion, filtersInsightConversion, getInsightToPropsConverter, getReactEmbeddingCodeGenerator, localeInsightConversion, singleAttributeBucketConversion, singleAttributeOrMeasureBucketConversion, sortsInsightConversion, } from "../../../utils/embeddingCodeGenerator";
import { chartAdditionalFactories, chartConfigInsightConversion } from "../chartCodeGenUtils";
var HeatmapDescriptor = /** @class */ (function (_super) {
    __extends(HeatmapDescriptor, _super);
    function HeatmapDescriptor() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.getEmbeddingCode = getReactEmbeddingCodeGenerator({
            component: {
                importType: "named",
                name: "Heatmap",
                package: "@gooddata/sdk-ui-charts",
            },
            insightToProps: getInsightToPropsConverter({
                measure: singleAttributeOrMeasureBucketConversion("measure", BucketNames.MEASURES),
                rows: singleAttributeBucketConversion("rows", BucketNames.VIEW),
                columns: singleAttributeBucketConversion("columns", BucketNames.STACK),
                filters: filtersInsightConversion("filters"),
                sortBy: sortsInsightConversion("sortBy"),
                config: chartConfigInsightConversion("config"),
                locale: localeInsightConversion("locale"),
                execConfig: executionConfigInsightConversion("execConfig"),
            }),
            additionalFactories: chartAdditionalFactories(),
        });
        return _this;
    }
    HeatmapDescriptor.prototype.getFactory = function () {
        return function (params) { return new PluggableHeatmap(params); };
    };
    HeatmapDescriptor.prototype.applyDrillDown = function (insight, drillDownContext, backendSupportsElementUris) {
        var withFilters = this.addFilters(insight, drillDownContext.drillDefinition, drillDownContext.event, backendSupportsElementUris);
        return modifyBucketsAttributesForDrillDown(withFilters, drillDownContext.drillDefinition);
    };
    HeatmapDescriptor.prototype.getMeta = function () {
        return {
            documentationUrl: "https://sdk.gooddata.com/gooddata-ui/docs/heatmap_component.html",
            supportsExport: true,
        };
    };
    HeatmapDescriptor.prototype.addFilters = function (source, drillConfig, event, backendSupportsElementUris) {
        var clicked = drillDownFromAttributeLocalId(drillConfig);
        var cutIntersection = (event.drillContext.intersection || []).filter(function (i) {
            return isDrillIntersectionAttributeItem(i.header) &&
                i.header.attributeHeader.localIdentifier === clicked;
        });
        return addIntersectionFiltersToInsight(source, cutIntersection, backendSupportsElementUris);
    };
    return HeatmapDescriptor;
}(BigChartDescriptor));
export { HeatmapDescriptor };
