// @gooddata/sdk-backend-tiger/esm/backend/organization/index.js
// (C) 2021-2022 GoodData Corporation
import { __awaiter, __generator } from "tslib";
import { SecuritySettingsService } from "./securitySettings";
import { OrganizationStylingService } from "./styling";
import { OrganizationSettingsService } from "./settings";
var TigerOrganization = /** @class */ (function () {
    function TigerOrganization(authCall, organizationId, organizationName) {
        this.authCall = authCall;
        this.organizationId = organizationId;
        this.organizationName = organizationName;
    }
    TigerOrganization.prototype.getDescriptor = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, organizationName, organizationId;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        // if we already have the data, no reason to download them again
                        if (this.organizationName) {
                            return [2 /*return*/, {
                                    id: this.organizationId,
                                    title: this.organizationName,
                                }];
                        }
                        return [4 /*yield*/, this.authCall(function (client) {
                                return client.profile.getCurrent();
                            })];
                    case 1:
                        _a = _b.sent(), organizationName = _a.organizationName, organizationId = _a.organizationId;
                        return [2 /*return*/, {
                                id: organizationId,
                                title: organizationName,
                            }];
                }
            });
        });
    };
    TigerOrganization.prototype.securitySettings = function () {
        return new SecuritySettingsService(this.organizationId);
    };
    TigerOrganization.prototype.styling = function () {
        return new OrganizationStylingService(this.authCall);
    };
    TigerOrganization.prototype.settings = function () {
        return new OrganizationSettingsService(this.authCall);
    };
    return TigerOrganization;
}());
export { TigerOrganization };
var TigerOrganizations = /** @class */ (function () {
    function TigerOrganizations(authCall) {
        this.authCall = authCall;
    }
    TigerOrganizations.prototype.getCurrentOrganization = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, organizationName, organizationId;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.authCall(function (client) {
                            return client.profile.getCurrent();
                        })];
                    case 1:
                        _a = _b.sent(), organizationName = _a.organizationName, organizationId = _a.organizationId;
                        return [2 /*return*/, new TigerOrganization(this.authCall, organizationId, organizationName)];
                }
            });
        });
    };
    return TigerOrganizations;
}());
export { TigerOrganizations };
