// @gooddata/sdk-backend-tiger/esm/convertors/fromBackend/analyticalDashboards/v1/AnalyticalDashboardConverter.js
import { __assign } from "tslib";
import { walkLayout } from "@gooddata/sdk-backend-spi";
import { idRef, } from "@gooddata/sdk-model";
import updateWith from "lodash/updateWith";
import { fixWidgetLegacyElementUris } from "../../fixLegacyElementUris";
import { cloneWithSanitizedIds } from "../../IdSanitization";
import { isInheritedObject } from "../../ObjectInheritance";
import { getShareStatus, stripQueryParams } from "../../utils";
import { convertDrillToCustomUrlInLayoutFromBackend } from "../DrillToCustomUrlConverter";
function setWidgetRefsInLayout(layout) {
    if (!layout) {
        return;
    }
    var widgetsPaths = [];
    walkLayout(layout, {
        widgetCallback: function (_, widgetPath) { return widgetsPaths.push(widgetPath); },
    });
    return widgetsPaths.reduce(function (layout, widgetPath, index) {
        return updateWith(layout, widgetPath, function (widget) {
            var temporaryWidgetId = widget.insight.identifier + "_widget-" + index;
            var convertedWidget = __assign(__assign({}, widget), { ref: idRef(temporaryWidgetId), uri: temporaryWidgetId, identifier: temporaryWidgetId });
            return fixWidgetLegacyElementUris(convertedWidget);
        });
    }, layout);
}
function getConvertedAnalyticalDashboardContent(analyticalDashboard) {
    return {
        dateFilterConfig: cloneWithSanitizedIds(analyticalDashboard.analyticalDashboard.dateFilterConfig),
        layout: convertDrillToCustomUrlInLayoutFromBackend(setWidgetRefsInLayout(cloneWithSanitizedIds(analyticalDashboard.analyticalDashboard.layout))),
    };
}
export function convertDashboard(analyticalDashboard, filterContext) {
    var _a, _b;
    var _c = analyticalDashboard.data, id = _c.id, _d = _c.attributes, attributes = _d === void 0 ? {} : _d, _e = _c.meta, meta = _e === void 0 ? {} : _e;
    var _f = attributes.title, title = _f === void 0 ? "" : _f, _g = attributes.description, description = _g === void 0 ? "" : _g, content = attributes.content;
    var isPrivate = (_b = (_a = meta.accessInfo) === null || _a === void 0 ? void 0 : _a.private) !== null && _b !== void 0 ? _b : false;
    var _h = getConvertedAnalyticalDashboardContent(content), dateFilterConfig = _h.dateFilterConfig, layout = _h.layout;
    return {
        type: "IDashboard",
        ref: idRef(id, "analyticalDashboard"),
        identifier: id,
        uri: stripQueryParams(analyticalDashboard.links.self),
        title: title,
        description: description,
        created: "",
        updated: "",
        // TODO: TIGER-HACK: inherited objects must be locked; they are read-only for all
        isLocked: isInheritedObject(analyticalDashboard.data),
        shareStatus: getShareStatus(isPrivate),
        isUnderStrictControl: true,
        tags: attributes.tags,
        filterContext: filterContext,
        dateFilterConfig: dateFilterConfig,
        layout: layout,
    };
}
export function convertFilterContextFromBackend(filterContext) {
    var _a = filterContext.data, id = _a.id, type = _a.type, attributes = _a.attributes;
    var _b = attributes, _c = _b.title, title = _c === void 0 ? "" : _c, _d = _b.description, description = _d === void 0 ? "" : _d, content = _b.content;
    return {
        ref: idRef(id, type),
        identifier: id,
        uri: filterContext.links.self,
        title: title,
        description: description,
        filters: convertFilterContextFilters(content),
    };
}
export function convertFilterContextFilters(content) {
    return cloneWithSanitizedIds(content.filterContext.filters);
}
