// @gooddata/sdk-model/esm/colors/index.js
// (C) 2019-2022 GoodData Corporation
import isEmpty from "lodash/isEmpty";
function isRgbChannel(c) {
    return c !== undefined && typeof c === "number" && c >= 0 && c < 256;
}
function isRgbColorValue(obj) {
    if (!isEmpty(obj)) {
        var _a = obj, r = _a.r, g = _a.g, b = _a.b;
        return isRgbChannel(r) && isRgbChannel(g) && isRgbChannel(b);
    }
    return false;
}
/**
 * Type guard checking whether the provided object is a {@link IColorPaletteItem}
 *
 * @public
 */
export function isColorPaletteItem(obj) {
    if (!isEmpty(obj)) {
        var _a = obj, guid = _a.guid, fill = _a.fill;
        return typeof guid === "string" && guid !== undefined && isRgbColorValue(fill);
    }
    return false;
}
//
// Type guards
//
/**
 * Type guard checking whether the provided object is a {@link IColorFromPalette}
 *
 * @public
 */
export function isColorFromPalette(obj) {
    return !isEmpty(obj) && obj.type === "guid";
}
/**
 * Type guard checking whether the provided object is a {@link IRgbColor}
 *
 * @public
 */
export function isRgbColor(obj) {
    return !isEmpty(obj) && obj.type === "rgb";
}
/**
 * Returns RGB code representing the color in the provided color palette items.
 *
 * @param item - color palette item
 * @returns an `rgb(red#,green#,blue#)` code
 * @public
 */
export function colorPaletteItemToRgb(item) {
    var _a = item.fill, r = _a.r, g = _a.g, b = _a.b;
    return "rgb(" + r + "," + g + "," + b + ")";
}
/**
 * Returns a list of RGB color codes for all items in the provided color palette.
 *
 * @param palette - color palette
 * @returns list with the same cardinality as the color palette. RGB colors appear in the same order in which
 * they appear in the palette
 * @public
 */
export function colorPaletteToColors(palette) {
    return palette.map(colorPaletteItemToRgb);
}
