// @gooddata/sdk-backend-tiger/esm/utils/errorHandling.js
// (C) 2019-2022 GoodData Corporation
import { isAnalyticalBackendError, NotAuthenticated, UnexpectedError, LimitReached, ContractExpired, } from "@gooddata/sdk-backend-spi";
export function convertApiError(error) {
    if (isAnalyticalBackendError(error)) {
        return error;
    }
    var notAuthenticated = createNotAuthenticatedError(error);
    if (notAuthenticated) {
        throw notAuthenticated;
    }
    var limitReached = createLimitReachedError(error);
    if (limitReached) {
        throw limitReached;
    }
    var contractExpired = createContractExpiredError(error);
    if (contractExpired) {
        return contractExpired;
    }
    return new UnexpectedError("An unexpected error has occurred", error);
}
export function createNotAuthenticatedError(error) {
    var axiosErrorResponse = error.response;
    if (!axiosErrorResponse || axiosErrorResponse.status !== 401) {
        return;
    }
    var exc = new NotAuthenticated("No session or session expired", error);
    // TODO: TIGER-HACK both of these params need to come from the backend.
    //  current problems:
    //  - some resources do not send login URL (empty 401), some do
    //  - no resources send returnRedirectParam
    exc.authenticationFlow = {
        loginUrl: "/appLogin",
        returnRedirectParam: "redirectTo",
    };
    return exc;
}
function createLimitReachedError(error) {
    var _a, _b;
    var axiosErrorResponse = error.response;
    if (!axiosErrorResponse ||
        axiosErrorResponse.status !== 400 ||
        !((_b = (_a = axiosErrorResponse.data) === null || _a === void 0 ? void 0 : _a.detail) === null || _b === void 0 ? void 0 : _b.includes("Reached plan limits"))) {
        return;
    }
    return new LimitReached("The limit reached. Upgrade your plan to create more objects.", error);
}
function createContractExpiredError(error) {
    var _a, _b, _c, _d;
    var axiosErrorResponse = error.response;
    if (!axiosErrorResponse ||
        axiosErrorResponse.status !== 403 ||
        (!((_b = (_a = axiosErrorResponse.data) === null || _a === void 0 ? void 0 : _a.detail) === null || _b === void 0 ? void 0 : _b.includes("Contract expired")) &&
            !((_d = (_c = axiosErrorResponse.data) === null || _c === void 0 ? void 0 : _c.detail) === null || _d === void 0 ? void 0 : _d.includes("Reason: EXPIRED")))) {
        return;
    }
    return new ContractExpired(axiosErrorResponse.data.tier || "unspecified", error);
}
