// @gooddata/sdk-backend-tiger/esm/backend/workspace/attributes/index.js
import { __awaiter, __generator } from "tslib";
// (C) 2019-2022 GoodData Corporation
import { NotSupported, UnexpectedResponseError, } from "@gooddata/sdk-backend-spi";
import { areObjRefsEqual, isIdentifierRef, } from "@gooddata/sdk-model";
import { TigerWorkspaceElements } from "./elements";
import { jsonApiHeaders, MetadataUtilities, JsonApiDatasetOutWithLinksTypeEnum, } from "@gooddata/api-client-tiger";
import flatMap from "lodash/flatMap";
import { invariant } from "ts-invariant";
import { convertAttributesWithSideloadedLabels, convertAttributeWithSideloadedLabels, convertDatasetWithLinks, } from "../../../convertors/fromBackend/MetadataConverter";
import { getIdOrigin } from "../../../convertors/fromBackend/ObjectInheritance";
var TigerWorkspaceAttributes = /** @class */ (function () {
    function TigerWorkspaceAttributes(authCall, workspace, dateFormatter) {
        var _this = this;
        this.authCall = authCall;
        this.workspace = workspace;
        this.dateFormatter = dateFormatter;
        this.getAttributeDisplayForm = function (ref) { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                        return [2 /*return*/, loadAttributeDisplayForm(client, this.workspace, ref)];
                    }); }); })];
            });
        }); };
        this.getAttribute = function (ref) { return __awaiter(_this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                        return [2 /*return*/, loadAttribute(client, this.workspace, ref)];
                    }); }); })];
            });
        }); };
    }
    TigerWorkspaceAttributes.prototype.elements = function () {
        return new TigerWorkspaceElements(this.authCall, this.workspace, this.dateFormatter);
    };
    TigerWorkspaceAttributes.prototype.getAttributeDisplayForms = function (refs) {
        var _this = this;
        return this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
            var allAttributes;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, loadAttributes(client, this.workspace)];
                    case 1:
                        allAttributes = _a.sent();
                        return [2 /*return*/, flatMap(allAttributes, function (attr) { return attr.displayForms; }).filter(function (df) {
                                return refs.find(function (ref) { return areObjRefsEqual(ref, df.ref); });
                            })];
                }
            });
        }); });
    };
    TigerWorkspaceAttributes.prototype.getAttributeByDisplayForm = function (ref) {
        var _this = this;
        return this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
            return [2 /*return*/, loadAttributeByDisplayForm(client, this.workspace, ref)];
        }); }); });
    };
    TigerWorkspaceAttributes.prototype.getAttributes = function (refs) {
        var _this = this;
        return this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
            var allAttributes;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, loadAttributes(client, this.workspace)];
                    case 1:
                        allAttributes = _a.sent();
                        return [2 /*return*/, allAttributes.filter(function (attr) { return refs.find(function (ref) { return areObjRefsEqual(ref, attr.ref); }); })];
                }
            });
        }); });
    };
    TigerWorkspaceAttributes.prototype.getCommonAttributes = function () {
        throw new NotSupported("not supported");
    };
    TigerWorkspaceAttributes.prototype.getCommonAttributesBatch = function () {
        throw new NotSupported("not supported");
    };
    TigerWorkspaceAttributes.prototype.getAttributeDatasetMeta = function (ref) {
        var _this = this;
        return this.authCall(function (client) {
            return loadAttributeDataset(client, _this.workspace, ref);
        });
    };
    return TigerWorkspaceAttributes;
}());
export { TigerWorkspaceAttributes };
function loadAttributeDisplayForm(client, workspaceId, ref) {
    return __awaiter(this, void 0, void 0, function () {
        var attributeRes, attributes, matchingLabel;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    invariant(isIdentifierRef(ref), "tiger backend only supports referencing by identifier");
                    return [4 /*yield*/, getAllEntitiesAttributesWithFilter(client, workspaceId, ref, [
                            "labels",
                            "defaultView",
                        ])];
                case 1:
                    attributeRes = _a.sent();
                    if (!attributeRes.data.data.length) {
                        throw new UnexpectedResponseError("The displayForm with id " + ref.identifier + " was not found", 404, attributeRes);
                    }
                    attributes = convertAttributesWithSideloadedLabels(attributeRes.data);
                    matchingLabel = findLabelInAttributes(attributes, ref);
                    invariant(matchingLabel, "inconsistent server response, RSQL matched but ref matching did not");
                    return [2 /*return*/, matchingLabel];
            }
        });
    });
}
function findLabelInAttributes(attributes, ref) {
    for (var _i = 0, attributes_1 = attributes; _i < attributes_1.length; _i++) {
        var attr = attributes_1[_i];
        for (var _a = 0, _b = attr.displayForms; _a < _b.length; _a++) {
            var df = _b[_a];
            if (areObjRefsEqual(df.ref, ref)) {
                return df;
            }
        }
    }
    return undefined;
}
function loadAttribute(client, workspaceId, ref) {
    invariant(isIdentifierRef(ref), "tiger backend only supports referencing by identifier");
    return client.entities
        .getEntityAttributes({
        workspaceId: workspaceId,
        objectId: ref.identifier,
        include: ["labels", "defaultView"],
    }, {
        headers: jsonApiHeaders,
    })
        .then(function (res) { return convertAttributeWithSideloadedLabels(res.data); });
}
function loadAttributeByDisplayForm(client, workspaceId, ref) {
    invariant(isIdentifierRef(ref), "tiger backend only supports referencing by identifier");
    return getAllEntitiesAttributesWithFilter(client, workspaceId, ref, ["labels"]).then(function (res) {
        var convertedAttributes = convertAttributesWithSideloadedLabels(res.data);
        var match = convertedAttributes.find(function (attr) {
            return attr.displayForms.some(function (df) { return df.id === ref.identifier; });
        });
        if (!match) {
            throw new UnexpectedResponseError("The displayForm with id " + ref.identifier + " was not found", 404, res);
        }
        return match;
    });
}
function loadAttributes(client, workspaceId) {
    return MetadataUtilities.getAllPagesOf(client, client.entities.getAllEntitiesAttributes, {
        workspaceId: workspaceId,
        include: ["labels"],
    })
        .then(MetadataUtilities.mergeEntitiesResults)
        .then(convertAttributesWithSideloadedLabels);
}
function loadAttributeDataset(client, workspace, ref) {
    invariant(isIdentifierRef(ref), "tiger backend only supports referencing by identifier");
    return client.entities
        .getEntityAttributes({
        workspaceId: workspace,
        objectId: ref.identifier,
        include: ["datasets"],
    }, {
        headers: jsonApiHeaders,
    })
        .then(function (res) {
        // if this happens then its either bad query parameterization or the backend is hosed badly
        invariant(res.data.included && res.data.included.length > 0, "server returned that attribute does not belong to any dataset");
        var datasets = res.data.included.filter(function (include) {
            return include.type === JsonApiDatasetOutWithLinksTypeEnum.DATASET;
        });
        return convertDatasetWithLinks(datasets[0]);
    });
}
function getAllEntitiesAttributesWithFilter(client, workspaceId, ref, includes) {
    invariant(isIdentifierRef(ref), "tiger backend only supports referencing by identifier");
    return client.entities.getAllEntitiesAttributes({
        workspaceId: workspaceId,
        // to be able to get the defaultView value, we need to load the attribute itself and then find the appropriate label inside it
        // otherwise, we would have to load the label first and then load its attribute to see the defaultView relation thus needing
        // an extra network request
        // tiger RSQL does not support prefixed ids, so we strip the prefix to load matches with or without prefix
        // and then find the prefixed value in the results
        filter: "labels.id==" + getIdOrigin(ref.identifier).id,
        include: includes,
    }, {
        headers: jsonApiHeaders,
    });
}
