// @gooddata/sdk-ui-filters/esm/AttributeFilter/Context/AttributeFilterComponentsContext.js
import { __rest } from "tslib";
// (C) 2022 GoodData Corporation
import React, { createContext, useContext } from "react";
import { ThrowMissingComponentError } from "../utils";
var AttributeFilterComponentsContext = createContext({
    ErrorComponent: ThrowMissingComponentError("ErrorComponent", "AttributeFilterComponentsContext"),
    LoadingComponent: ThrowMissingComponentError("LoadingComponent", "AttributeFilterComponentsContext"),
    DropdownButtonComponent: ThrowMissingComponentError("DropdownButtonComponent", "AttributeFilterComponentsContext"),
    DropdownBodyComponent: ThrowMissingComponentError("DropdownBodyComponent", "AttributeFilterComponentsContext"),
    DropdownActionsComponent: ThrowMissingComponentError("DropdownActionsComponent", "AttributeFilterComponentsContext"),
    ElementsSearchBarComponent: ThrowMissingComponentError("ElementsSearchBarComponent", "AttributeFilterComponentsContext"),
    ElementsSelectComponent: ThrowMissingComponentError("ElementsSelectComponent", "AttributeFilterComponentsContext"),
    ElementsSelectLoadingComponent: ThrowMissingComponentError("ElementsSelectLoadingComponent", "AttributeFilterComponentsContext"),
    ElementsSelectItemComponent: ThrowMissingComponentError("ElementsSelectItemComponent", "AttributeFilterComponentsContext"),
    ElementsSelectErrorComponent: ThrowMissingComponentError("ElementsSelectErrorComponent", "AttributeFilterComponentsContext"),
    ElementsSelectActionsComponent: ThrowMissingComponentError("ElementsSelectActionsComponent", "AttributeFilterComponentsContext"),
    EmptyResultComponent: ThrowMissingComponentError("EmptyResultComponent", "AttributeFilterComponentsContext"),
    StatusBarComponent: ThrowMissingComponentError("StatusBarComponent", "AttributeFilterComponentsContext"),
    FilterError: ThrowMissingComponentError("FilterError", "AttributeFilterComponentsContext"),
    FilterLoading: ThrowMissingComponentError("FilterLoading", "AttributeFilterComponentsContext"),
});
AttributeFilterComponentsContext.displayName = "AttributeFilterComponentsContext";
/**
 * @internal
 */
export var useAttributeFilterComponentsContext = function () {
    return useContext(AttributeFilterComponentsContext);
};
/**
 * @internal
 */
export var AttributeFilterComponentsProvider = function (props) {
    var children = props.children, components = __rest(props, ["children"]);
    return (React.createElement(AttributeFilterComponentsContext.Provider, { value: components }, children));
};
