// @gooddata/sdk-ui-dashboard/esm/model/store/ui/uiReducers.js
var openScheduleEmailDialog = function (state) {
    state.scheduleEmailDialog.open = true;
};
var closeScheduleEmailDialog = function (state) {
    state.scheduleEmailDialog.open = false;
};
var setScheduleEmailDialogDefaultAttachment = function (state, action) {
    state.scheduleEmailDialog.defaultAttachmentRef = action.payload;
};
var resetScheduleEmailDialogDefaultAttachment = function (state) {
    state.scheduleEmailDialog.defaultAttachmentRef = undefined;
};
var openScheduleEmailManagementDialog = function (state) {
    state.scheduleEmailManagementDialog.open = true;
};
var closeScheduleEmailManagementDialog = function (state) {
    state.scheduleEmailManagementDialog.open = false;
};
var openSaveAsDialog = function (state) {
    state.saveAsDialog.open = true;
};
var closeSaveAsDialog = function (state) {
    state.saveAsDialog.open = false;
};
var openShareDialog = function (state) {
    state.shareDialog.open = true;
};
var closeShareDialog = function (state) {
    state.shareDialog.open = false;
};
var openDeleteDialog = function (state) {
    state.deleteDialog.open = true;
};
var closeDeleteDialog = function (state) {
    state.deleteDialog.open = false;
};
var setFilterBarHeight = function (state, action) {
    state.filterBar.height = action.payload;
};
var setFilterBarExpanded = function (state, action) {
    state.filterBar.expanded = action.payload;
};
var closeKpiAlertDialog = function (state) {
    state.kpiAlerts.openedWidgetRef = undefined;
};
var openKpiAlertDialog = function (state, action) {
    state.kpiAlerts.openedWidgetRef = action.payload;
};
var highlightKpiAlert = function (state, action) {
    state.kpiAlerts.highlightedWidgetRef = action.payload;
};
var setMenuButtonItemsVisibility = function (state, action) {
    state.menuButton.itemsVisibility = action.payload;
};
var setRenderMode = function (state, action) {
    state.renderMode = action.payload;
};
var setEditRenderMode = function (state) {
    state.renderMode = "edit";
};
var setViewRenderMode = function (state) {
    state.renderMode = "view";
};
var setActiveHeaderIndex = function (state, action) {
    state.activeHeaderIndex = action.payload;
};
var selectWidget = function (state, action) {
    state.selectedWidgetRef = action.payload;
};
var clearWidgetSelection = function (state) {
    state.selectedWidgetRef = undefined;
};
export var uiReducers = {
    openScheduleEmailDialog: openScheduleEmailDialog,
    closeScheduleEmailDialog: closeScheduleEmailDialog,
    setScheduleEmailDialogDefaultAttachment: setScheduleEmailDialogDefaultAttachment,
    resetScheduleEmailDialogDefaultAttachment: resetScheduleEmailDialogDefaultAttachment,
    openScheduleEmailManagementDialog: openScheduleEmailManagementDialog,
    closeScheduleEmailManagementDialog: closeScheduleEmailManagementDialog,
    openSaveAsDialog: openSaveAsDialog,
    closeSaveAsDialog: closeSaveAsDialog,
    setFilterBarHeight: setFilterBarHeight,
    setFilterBarExpanded: setFilterBarExpanded,
    closeKpiAlertDialog: closeKpiAlertDialog,
    openKpiAlertDialog: openKpiAlertDialog,
    highlightKpiAlert: highlightKpiAlert,
    openShareDialog: openShareDialog,
    closeShareDialog: closeShareDialog,
    openDeleteDialog: openDeleteDialog,
    closeDeleteDialog: closeDeleteDialog,
    setMenuButtonItemsVisibility: setMenuButtonItemsVisibility,
    setRenderMode: setRenderMode,
    setEditRenderMode: setEditRenderMode,
    setViewRenderMode: setViewRenderMode,
    setActiveHeaderIndex: setActiveHeaderIndex,
    selectWidget: selectWidget,
    clearWidgetSelection: clearWidgetSelection,
};
