// @gooddata/api-client-tiger/esm/gd-tiger-model/typeGuards.js
// (C) 2019-2022 GoodData Corporation
import { JsonApiDashboardPluginOutWithLinksTypeEnum, JsonApiFilterContextInTypeEnum, JsonApiVisualizationObjectOutWithLinksTypeEnum, } from "../generated/metadata-json-api";
export function isAttributeHeader(header) {
    return header && header.attributeHeader !== undefined;
}
export var isAfmObjectIdentifier = function (value) {
    var _a, _b, _c, _d;
    return !!(((_b = (_a = value) === null || _a === void 0 ? void 0 : _a.identifier) === null || _b === void 0 ? void 0 : _b.id) && ((_d = (_c = value) === null || _c === void 0 ? void 0 : _c.identifier) === null || _d === void 0 ? void 0 : _d.type));
};
export function isResultAttributeHeader(header) {
    return header.attributeHeader !== undefined;
}
export function isResultMeasureHeader(header) {
    return header.measureHeader !== undefined;
}
export function isResultTotalHeader(header) {
    return header.totalHeader !== undefined;
}
export function isVisualizationObjectsItem(visualizationObject) {
    return (visualizationObject.type ===
        JsonApiVisualizationObjectOutWithLinksTypeEnum.VISUALIZATION_OBJECT);
}
export function isFilterContextData(filterContext) {
    return filterContext.type === JsonApiFilterContextInTypeEnum.FILTER_CONTEXT;
}
export function isDashboardPluginsItem(dashboardPlugin) {
    return (dashboardPlugin.type ===
        JsonApiDashboardPluginOutWithLinksTypeEnum.DASHBOARD_PLUGIN);
}
