// @gooddata/sdk-backend-spi/esm/deprecated/ldm/metadata/index.js
// (C) 2019-2022 GoodData Corporation
/* eslint-disable @typescript-eslint/no-empty-interface, import/namespace */
import * as m from "@gooddata/sdk-model";
import { isAttributeMetadataObject } from "./attribute";
import { isAttributeDisplayFormMetadataObject, } from "./attributeDisplayForm";
import { isDashboardMetadataObject } from "./dashboard";
import { isDataSetMetadataObject } from "./dataSet";
import { isFactMetadataObject } from "./fact";
import { isMeasureMetadataObject, isMeasureMetadataObjectDefinition, } from "./measure";
import { isMetadataObject } from "./types";
import { isVariableMetadataObject } from "./variable";
export { isMetadataObject, isAttributeMetadataObject, isAttributeDisplayFormMetadataObject, isFactMetadataObject, isMeasureMetadataObject, isMeasureMetadataObjectDefinition, isDataSetMetadataObject, isVariableMetadataObject, isDashboardMetadataObject, };
/**
 * Get metadata object identifier
 * @deprecated Use {@link @gooddata/sdk-model#metadataObjectId}
 * @public
 */
export var metadataObjectId = m.metadataObjectId;
