// @gooddata/sdk-ui-dashboard/esm/presentation/scheduledEmail/DefaultScheduledEmailDialog/ScheduledMailDialogRenderer/RepeatSelect/RepeatExecuteOnSelect.js
// (C) 2019-2022 GoodData Corporation
import React, { useMemo } from "react";
import { useIntl } from "react-intl";
import { Dropdown, DropdownList, DropdownButton, SingleSelectListItem } from "@gooddata/sdk-ui-kit";
import invariant from "ts-invariant";
import { DEFAULT_DROPDOWN_ALIGN_POINTS, DEFAULT_DROPDOWN_ZINDEX, REPEAT_EXECUTE_ON } from "../../constants";
import { getDate, getIntlDayName, getWeek } from "../../utils/datetime";
import { messages } from "../../../../../locales";
var DROPDOWN_WIDTH = 154;
export var RepeatExecuteOnSelect = function (props) {
    var onChange = props.onChange, repeatExecuteOn = props.repeatExecuteOn, startDate = props.startDate;
    var intl = useIntl();
    var repeatExecuteOnItems = useMemo(function () {
        return [REPEAT_EXECUTE_ON.DAY_OF_MONTH, REPEAT_EXECUTE_ON.DAY_OF_WEEK].map(function (id) {
            return {
                id: id,
                title: intl.formatMessage(messages["scheduleDialogEmailRepeatsExecuteOn_" + id], {
                    date: getDate(startDate),
                    day: getIntlDayName(intl, startDate),
                    week: getWeek(startDate),
                }),
            };
        });
    }, [intl, startDate]);
    var repeatExecuteOnItem = repeatExecuteOnItems.find(function (item) { return item.id === repeatExecuteOn; });
    invariant(repeatExecuteOnItem, "Inconsistent state in RepeatExecuteOnSelect");
    return (React.createElement(Dropdown, { alignPoints: DEFAULT_DROPDOWN_ALIGN_POINTS, className: "gd-schedule-email-dialog-repeat-execute-on s-gd-schedule-email-dialog-repeat-execute-on", renderBody: function (_a) {
            var closeDropdown = _a.closeDropdown, isMobile = _a.isMobile;
            return (React.createElement(DropdownList, { width: DROPDOWN_WIDTH, items: repeatExecuteOnItems, isMobile: isMobile, renderItem: function (_a) {
                    var item = _a.item;
                    return (React.createElement(SingleSelectListItem, { title: item.title, onClick: function () {
                            onChange(item.id);
                            closeDropdown();
                        }, isSelected: repeatExecuteOnItem.id === item.id }));
                } }));
        }, renderButton: function (_a) {
            var toggleDropdown = _a.toggleDropdown;
            return (React.createElement(DropdownButton, { value: repeatExecuteOnItem.title, onClick: toggleDropdown }));
        }, overlayPositionType: "sameAsTarget", overlayZIndex: DEFAULT_DROPDOWN_ZINDEX }));
};
