// @gooddata/sdk-ui/esm/base/localization/TranslationsProvider.js
import { __extends } from "tslib";
// (C) 2007-2022 GoodData Corporation
import React from "react";
import { injectIntl } from "react-intl";
import { messages } from "../../locales";
import { emptyHeaderTitleFromIntl } from "./intlUtils";
var getNumericSymbols = function (intl) {
    return [messages.k, messages.m, messages.g, messages.t, messages.p, messages.e].map(function (desc) {
        return intl.formatMessage(desc);
    });
};
/**
 * @internal
 */
var TranslationsProvider = /** @class */ (function (_super) {
    __extends(TranslationsProvider, _super);
    function TranslationsProvider() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TranslationsProvider.prototype.render = function () {
        var props = {
            numericSymbols: getNumericSymbols(this.props.intl),
            emptyHeaderString: emptyHeaderTitleFromIntl(this.props.intl),
            intl: this.props.intl,
        };
        return this.props.children(props);
    };
    return TranslationsProvider;
}(React.PureComponent));
export { TranslationsProvider };
/**
 * @internal
 */
export var IntlTranslationsProvider = injectIntl(TranslationsProvider);
