// @gooddata/sdk-ui-dashboard/esm/presentation/scheduledEmail/DefaultScheduledEmailDialog/index.js
// (C) 2019-2022 GoodData Corporation
import React from "react";
import invariant from "ts-invariant";
import { ScheduledMailDialogRenderer } from "./ScheduledMailDialogRenderer/ScheduledMailDialogRenderer";
import { useScheduledEmail } from "./useScheduledEmail";
/**
 * @alpha
 */
export var DefaultScheduledEmailDialog = function (props) {
    var onSubmit = props.onSubmit, onSuccess = props.onSuccess, onError = props.onError, onSave = props.onSave, onSaveSuccess = props.onSaveSuccess, onSaveError = props.onSaveError, onCancel = props.onCancel, isVisible = props.isVisible, editSchedule = props.editSchedule;
    var _a = useScheduledEmail({
        onSubmit: onSubmit,
        onSubmitSuccess: onSuccess,
        onSubmitError: onError,
        onSave: onSave,
        onSaveSuccess: onSaveSuccess,
        onSaveError: onSaveError,
    }), currentUser = _a.currentUser, dashboardRef = _a.dashboardRef, dashboardTitle = _a.dashboardTitle, dashboardInsightWidgets = _a.dashboardInsightWidgets, hasDefaultFilters = _a.hasDefaultFilters, handleCreateScheduledEmail = _a.handleCreateScheduledEmail, handleSaveScheduledEmail = _a.handleSaveScheduledEmail, locale = _a.locale, canListUsersInWorkspace = _a.canListUsersInWorkspace, canExportReport = _a.canExportReport, dateFormat = _a.dateFormat, enableKPIDashboardSchedule = _a.enableKPIDashboardSchedule, enableKPIDashboardScheduleRecipients = _a.enableKPIDashboardScheduleRecipients, enableWidgetExportScheduling = _a.enableWidgetExportScheduling, defaultAttachment = _a.defaultAttachment;
    // trigger the invariant only if the user tries to open the dialog
    if (isVisible) {
        invariant(enableKPIDashboardSchedule, "Feature flag enableKPIDashboardSchedule must be enabled to make ScheduledMailDialog work properly.");
    }
    if (!isVisible) {
        return null;
    }
    return (React.createElement(ScheduledMailDialogRenderer, { editSchedule: editSchedule, locale: locale, canListUsersInProject: canListUsersInWorkspace, canExportReport: canExportReport, enableKPIDashboardScheduleRecipients: enableKPIDashboardScheduleRecipients, enableWidgetExportScheduling: enableWidgetExportScheduling, dateFormat: dateFormat, currentUser: currentUser, dashboard: dashboardRef, dashboardTitle: dashboardTitle, dashboardInsightWidgets: dashboardInsightWidgets, hasDefaultFilters: hasDefaultFilters, onSubmit: handleCreateScheduledEmail, onSave: handleSaveScheduledEmail, onCancel: onCancel, onError: onError, defaultAttachment: defaultAttachment }));
};
