// @gooddata/sdk-ui-ext/esm/internal/components/configurationControls/DataLabelsControl.js
import { __extends } from "tslib";
// (C) 2019-2022 GoodData Corporation
import React from "react";
import { injectIntl } from "react-intl";
import DropdownControl from "./DropdownControl";
import { dataLabelsDropdownItems } from "../../constants/dropdowns";
import { getTranslatedDropdownItems } from "../../utils/translations";
import { messages } from "../../../locales";
var DataLabelsControl = /** @class */ (function (_super) {
    __extends(DataLabelsControl, _super);
    function DataLabelsControl() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    DataLabelsControl.prototype.render = function () {
        var _a, _b, _c, _d, _e, _f;
        var _g = this.props, pushData = _g.pushData, properties = _g.properties, intl = _g.intl, isDisabled = _g.isDisabled, showDisabledMessage = _g.showDisabledMessage, defaultValue = _g.defaultValue, isTotalsDisabled = _g.isTotalsDisabled, enableSeparateTotalLabels = _g.enableSeparateTotalLabels;
        var dataLabels = (_c = (_b = (_a = properties === null || properties === void 0 ? void 0 : properties.controls) === null || _a === void 0 ? void 0 : _a.dataLabels) === null || _b === void 0 ? void 0 : _b.visible) !== null && _c !== void 0 ? _c : defaultValue;
        var totalLabels = (_f = (_e = (_d = properties === null || properties === void 0 ? void 0 : properties.controls) === null || _d === void 0 ? void 0 : _d.dataLabels) === null || _e === void 0 ? void 0 : _e.totalsVisible) !== null && _f !== void 0 ? _f : defaultValue;
        return (React.createElement("div", { className: "s-data-labels-config" },
            React.createElement(DropdownControl, { value: dataLabels, valuePath: "dataLabels.visible", labelText: messages.dataLabels.id, disabled: isDisabled, properties: properties, pushData: pushData, items: getTranslatedDropdownItems(dataLabelsDropdownItems, intl), showDisabledMessage: showDisabledMessage }),
            enableSeparateTotalLabels && (React.createElement(DropdownControl, { value: totalLabels, valuePath: "dataLabels.totalsVisible", labelText: messages.totalLabels.id, disabled: isTotalsDisabled, properties: properties, pushData: pushData, items: getTranslatedDropdownItems(dataLabelsDropdownItems, intl), showDisabledMessage: showDisabledMessage }))));
    };
    DataLabelsControl.defaultProps = {
        defaultValue: "auto",
        showDisabledMessage: false,
        isTotalsDisabled: true,
        enableSeparateTotalLabels: false,
    };
    return DataLabelsControl;
}(React.Component));
export default injectIntl(DataLabelsControl);
