// @gooddata/sdk-backend-bear/esm/backend/workspace/catalog/catalog.js
// (C) 2019-2022 GoodData Corporation
import invariant from "ts-invariant";
import { isCatalogAttribute, isCatalogFact, isCatalogMeasure, isCatalogDateDataset, } from "@gooddata/sdk-model";
import { BearWorkspaceCatalogAvailableItemsFactory } from "./availableItemsFactory";
var BearWorkspaceCatalog = /** @class */ (function () {
    function BearWorkspaceCatalog(authCall, workspace, catalogGroups, items, options, mappings) {
        this.authCall = authCall;
        this.workspace = workspace;
        this.catalogGroups = catalogGroups;
        this.items = items;
        this.options = options;
        this.mappings = mappings;
    }
    BearWorkspaceCatalog.prototype.groups = function () {
        invariant(this.options.loadGroups, "Catalog groups are not loaded.");
        return this.catalogGroups;
    };
    BearWorkspaceCatalog.prototype.allItems = function () {
        return this.items;
    };
    BearWorkspaceCatalog.prototype.attributes = function () {
        return this.items.filter(isCatalogAttribute);
    };
    BearWorkspaceCatalog.prototype.measures = function () {
        return this.items.filter(isCatalogMeasure);
    };
    BearWorkspaceCatalog.prototype.facts = function () {
        return this.items.filter(isCatalogFact);
    };
    BearWorkspaceCatalog.prototype.dateDatasets = function () {
        return this.items.filter(isCatalogDateDataset);
    };
    BearWorkspaceCatalog.prototype.availableItems = function () {
        return new BearWorkspaceCatalogAvailableItemsFactory(this.authCall, this.workspace, this.catalogGroups, this.items, this.options, this.mappings);
    };
    return BearWorkspaceCatalog;
}());
export { BearWorkspaceCatalog };
