// @gooddata/sdk-backend-bear/esm/backend/workspace/catalog/catalogWithAvailableItems.js
// (C) 2019-2022 GoodData Corporation
import invariant from "ts-invariant";
import { isCatalogAttribute, isCatalogFact, isCatalogMeasure, isCatalogDateDataset, } from "@gooddata/sdk-model";
var BearWorkspaceCatalogWithAvailableItems = /** @class */ (function () {
    function BearWorkspaceCatalogWithAvailableItems(catalogGroups, items, availableItems, options) {
        this.catalogGroups = catalogGroups;
        this.items = items;
        this.availableItems = availableItems;
        this.options = options;
    }
    BearWorkspaceCatalogWithAvailableItems.prototype.groups = function () {
        invariant(this.options.loadGroups, "Catalog groups are not loaded.");
        return this.catalogGroups;
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.allItems = function () {
        return this.items;
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.attributes = function () {
        return this.items.filter(isCatalogAttribute);
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.measures = function () {
        return this.items.filter(isCatalogMeasure);
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.facts = function () {
        return this.items.filter(isCatalogFact);
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.dateDatasets = function () {
        return this.items.filter(isCatalogDateDataset);
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.allAvailableItems = function () {
        return this.availableItems;
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.availableAttributes = function () {
        return this.availableItems.filter(isCatalogAttribute);
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.availableMeasures = function () {
        return this.availableItems.filter(isCatalogMeasure);
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.availableFacts = function () {
        return this.availableItems.filter(isCatalogFact);
    };
    BearWorkspaceCatalogWithAvailableItems.prototype.availableDateDatasets = function () {
        return this.availableItems.filter(isCatalogDateDataset);
    };
    return BearWorkspaceCatalogWithAvailableItems;
}());
export { BearWorkspaceCatalogWithAvailableItems };
