// @gooddata/sdk-backend-tiger/esm/convertors/fromBackend/AccessControlConverter.js
// (C) 2022-2023 GoodData Corporation
import { idRef, } from "@gooddata/sdk-model";
var getPermissionLevels = function (permissions, source) {
    if (permissions === void 0) { permissions = []; }
    return permissions
        .filter(function (permission) { return permission.source === source; })
        .map(function (permission) { return permission.level; });
};
export var convertUserPermission = function (user) {
    var _a, _b, _c, _d;
    return ({
        type: "granularUser",
        user: {
            ref: idRef(user.id),
            uri: user.id,
            email: (_a = user.email) !== null && _a !== void 0 ? _a : user.id,
            login: (_b = user.email) !== null && _b !== void 0 ? _b : user.id,
            fullName: (_d = (_c = user.name) !== null && _c !== void 0 ? _c : user.email) !== null && _d !== void 0 ? _d : user.id,
        },
        permissions: getPermissionLevels(user.permissions, "direct"),
        inheritedPermissions: getPermissionLevels(user.permissions, "indirect"),
    });
};
export var convertUserGroupPermission = function (group) {
    var _a;
    return ({
        type: "granularGroup",
        userGroup: { ref: idRef(group.id), name: (_a = group.name) !== null && _a !== void 0 ? _a : group.id },
        permissions: getPermissionLevels(group.permissions, "direct"),
        inheritedPermissions: getPermissionLevels(group.permissions, "indirect"),
    });
};
export var convertUserAssignee = function (user) {
    var _a, _b;
    return ({
        type: "user",
        ref: idRef(user.id),
        name: (_b = (_a = user.name) !== null && _a !== void 0 ? _a : user.email) !== null && _b !== void 0 ? _b : user.id,
        status: "ENABLED",
    });
};
export var convertUserGroupAssignee = function (group) {
    var _a;
    return ({
        type: "group",
        ref: idRef(group.id),
        name: (_a = group.name) !== null && _a !== void 0 ? _a : group.id,
    });
};
