// @gooddata/sdk-backend-tiger/esm/convertors/fromBackend/ObjectInheritance.js
// (C) 2020-2023 GoodData Corporation
import last from "lodash/last";
import first from "lodash/first";
var PrefixSeparator = ":";
export function isInheritedObject(obj) {
    var originType = getObjectOrigin(obj).originType;
    return originType === "PARENT";
}
export function getObjectOrigin(obj) {
    var origin = (obj.meta || {}).origin;
    return origin || { originType: "NATIVE", originId: "" };
}
/**
 * This method split id by Prefix separator (:) and return origin info
 *
 * @remarks
 * Id without prefix - LOCAL origin type with not origin id
 * Id with prefix - REMOTE origin type with origin id as first part of id (before :) and
 *  id as second part if id (after :)
 *
 * @param id - string that represent id with or without prefix
 * @internal
 */
export function getIdOrigin(id) {
    var data = id.split(PrefixSeparator);
    //prefix + id
    if (data.length === 2) {
        return {
            originType: "PARENT",
            originId: first(data),
            id: last(data),
        };
    }
    return {
        originType: "NATIVE",
        originId: "",
        id: id,
    };
}
