// @gooddata/sdk-ui-filters/esm/AttributeFilterHandler/internal/redux/loadCustomElements/loadCustomElementsReducers.js
import identity from "lodash/identity";
import { getElementCacheKey } from "../common/selectors";
var loadCustomElementsRequest = identity;
var loadCustomElementsStart = identity;
var loadCustomElementsSuccess = function (state, action) {
    action.payload.elements.forEach(function (el) {
        var cacheKey = getElementCacheKey(state, el);
        if (!state.elements.cache[cacheKey]) {
            state.elements.cache[cacheKey] = el;
        }
    });
};
var loadCustomElementsError = identity;
var loadCustomElementsCancelRequest = identity;
var loadCustomElementsCancel = identity;
/**
 * @internal
 */
export var loadCustomElementsReducers = {
    loadCustomElementsRequest: loadCustomElementsRequest,
    loadCustomElementsStart: loadCustomElementsStart,
    loadCustomElementsSuccess: loadCustomElementsSuccess,
    loadCustomElementsError: loadCustomElementsError,
    loadCustomElementsCancelRequest: loadCustomElementsCancelRequest,
    loadCustomElementsCancel: loadCustomElementsCancel,
};
