// @gooddata/api-client-tiger/esm/client.js
// (C) 2019-2023 GoodData Corporation
import { tigerExecutionClientFactory } from "./execution";
import { tigerExecutionResultClientFactory } from "./executionResult";
import { LabelElementsConfiguration, LabelElementsBaseApi, tigerLabelElementsClientFactory, } from "./labelElements";
import { tigerValidObjectsClientFactory } from "./validObjects";
import { setAxiosAuthorizationToken } from "./axios";
import { tigerLayoutClientFactory } from "./layout";
import { tigerAfmExplainClientFactory } from "./explain";
import { tigerActionsClientFactory } from "./actions";
import { tigerAuthActionsClientFactory } from "./authActions";
import { MetadataConfiguration, MetadataBaseApi, tigerEntitiesObjectsClientFactory, } from "./entitiesObjects";
import { tigerProfileClientFactory, } from "./profile";
import { tigerExportClientFactory, } from "./export";
import { ScanModelConfiguration, ScanModelBaseApi, tigerScanModelClientFactory, } from "./scanModel";
export { tigerExecutionClientFactory, tigerEntitiesObjectsClientFactory, tigerExecutionResultClientFactory, tigerLabelElementsClientFactory, tigerValidObjectsClientFactory, tigerLayoutClientFactory, tigerAfmExplainClientFactory, tigerProfileClientFactory, tigerActionsClientFactory, tigerAuthActionsClientFactory, tigerScanModelClientFactory, tigerExportClientFactory, MetadataConfiguration, MetadataBaseApi, LabelElementsConfiguration, LabelElementsBaseApi, ScanModelConfiguration, ScanModelBaseApi, };
/**
 * Tiger execution client
 *
 */
export var tigerClientFactory = function (axios) {
    var execution = tigerExecutionClientFactory(axios);
    var executionResult = tigerExecutionResultClientFactory(axios);
    var labelElements = tigerLabelElementsClientFactory(axios);
    var validObjects = tigerValidObjectsClientFactory(axios);
    var declarativeLayout = tigerLayoutClientFactory(axios);
    var explain = tigerAfmExplainClientFactory(axios);
    var entities = tigerEntitiesObjectsClientFactory(axios);
    var profile = tigerProfileClientFactory(axios);
    var actions = tigerActionsClientFactory(axios);
    var authActions = tigerAuthActionsClientFactory(axios);
    var scanModel = tigerScanModelClientFactory(axios);
    var exportFactory = tigerExportClientFactory(axios);
    return {
        axios: axios,
        execution: execution,
        executionResult: executionResult,
        labelElements: labelElements,
        validObjects: validObjects,
        declarativeLayout: declarativeLayout,
        explain: explain,
        entities: entities,
        profile: profile,
        actions: actions,
        authActions: authActions,
        scanModel: scanModel,
        setApiToken: function (token) {
            setAxiosAuthorizationToken(axios, token);
        },
        export: exportFactory,
    };
};
