// @gooddata/sdk-backend-tiger/esm/convertors/fromBackend/analyticalDashboards/v2/AnalyticalDashboardConverter.js
import { __assign } from "tslib";
import { walkLayout } from "@gooddata/sdk-backend-spi";
import { idRef, } from "@gooddata/sdk-model";
import updateWith from "lodash/updateWith";
import { cloneWithSanitizedIds } from "../../IdSanitization";
import { isInheritedObject } from "../../ObjectInheritance";
import { fixWidgetLegacyElementUris } from "../../fixLegacyElementUris";
import { convertDrillToCustomUrlInLayoutFromBackend } from "../DrillToCustomUrlConverter";
import { getShareStatus, stripQueryParams } from "../../utils";
function setWidgetRefsInLayout(layout) {
    if (!layout) {
        return;
    }
    var widgetsPaths = [];
    walkLayout(layout, {
        widgetCallback: function (_, widgetPath) { return widgetsPaths.push(widgetPath); },
    });
    return widgetsPaths.reduce(function (layout, widgetPath, index) {
        return updateWith(layout, widgetPath, function (widget) {
            var temporaryWidgetId = widget.insight.identifier + "_widget-" + index;
            var convertedWidget = __assign(__assign({}, widget), { ref: idRef(temporaryWidgetId), uri: temporaryWidgetId, identifier: temporaryWidgetId });
            return fixWidgetLegacyElementUris(convertedWidget);
        });
    }, layout);
}
function convertDashboardPluginLink(pluginLink) {
    return {
        type: "IDashboardPluginLink",
        plugin: cloneWithSanitizedIds(pluginLink.plugin),
        parameters: pluginLink.parameters,
    };
}
function getConvertedAnalyticalDashboardContent(analyticalDashboard) {
    var _a;
    return {
        dateFilterConfig: cloneWithSanitizedIds(analyticalDashboard.dateFilterConfig),
        layout: convertDrillToCustomUrlInLayoutFromBackend(setWidgetRefsInLayout(cloneWithSanitizedIds(analyticalDashboard.layout))),
        plugins: (_a = analyticalDashboard.plugins) === null || _a === void 0 ? void 0 : _a.map(convertDashboardPluginLink),
    };
}
export function convertDashboard(analyticalDashboard, filterContext) {
    var _a, _b, _c;
    var _d = analyticalDashboard.data, id = _d.id, _e = _d.attributes, attributes = _e === void 0 ? {} : _e, _f = _d.meta, meta = _f === void 0 ? {} : _f;
    var _g = attributes.title, title = _g === void 0 ? "" : _g, _h = attributes.description, description = _h === void 0 ? "" : _h, content = attributes.content;
    var isPrivate = (_b = (_a = meta.accessInfo) === null || _a === void 0 ? void 0 : _a.private) !== null && _b !== void 0 ? _b : false;
    var _j = getConvertedAnalyticalDashboardContent(content), dateFilterConfig = _j.dateFilterConfig, layout = _j.layout, plugins = _j.plugins;
    return {
        type: "IDashboard",
        ref: idRef(id, "analyticalDashboard"),
        identifier: id,
        uri: stripQueryParams(analyticalDashboard.links.self),
        title: title,
        description: description,
        created: "",
        updated: "",
        // TODO: TIGER-HACK: inherited objects must be locked; they are read-only for all
        isLocked: isInheritedObject(analyticalDashboard.data),
        shareStatus: getShareStatus(isPrivate),
        isUnderStrictControl: true,
        tags: (_c = attributes.tags) !== null && _c !== void 0 ? _c : [],
        filterContext: filterContext,
        dateFilterConfig: dateFilterConfig,
        layout: layout,
        plugins: plugins,
    };
}
export function convertFilterContextFromBackend(filterContext) {
    var _a = filterContext.data, id = _a.id, type = _a.type, attributes = _a.attributes;
    var _b = attributes, _c = _b.title, title = _c === void 0 ? "" : _c, _d = _b.description, description = _d === void 0 ? "" : _d, content = _b.content;
    return {
        ref: idRef(id, type),
        identifier: id,
        uri: filterContext.links.self,
        title: title,
        description: description,
        filters: convertFilterContextFilters(content),
    };
}
export function convertFilterContextFilters(content) {
    return cloneWithSanitizedIds(content.filters);
}
export function convertDashboardPlugin(plugin) {
    var _a = plugin.data, id = _a.id, type = _a.type, attributes = _a.attributes;
    var _b = attributes, _c = _b.title, title = _c === void 0 ? "" : _c, _d = _b.description, description = _d === void 0 ? "" : _d, content = _b.content, tags = _b.tags;
    var url = content.url;
    return {
        ref: idRef(id, type),
        identifier: id,
        uri: plugin.links.self,
        name: title,
        description: description,
        tags: tags !== null && tags !== void 0 ? tags : [],
        type: "IDashboardPlugin",
        url: url,
    };
}
export function convertDashboardPluginWithLinks(plugin) {
    var id = plugin.id, type = plugin.type, attributes = plugin.attributes;
    var _a = attributes, _b = _a.title, title = _b === void 0 ? "" : _b, _c = _a.description, description = _c === void 0 ? "" : _c, content = _a.content, tags = _a.tags;
    var url = content.url;
    return {
        ref: idRef(id, type),
        identifier: id,
        uri: plugin.links.self,
        name: title,
        description: description,
        tags: tags !== null && tags !== void 0 ? tags : [],
        type: "IDashboardPlugin",
        url: url,
    };
}
