// @gooddata/sdk-backend-bear/esm/convertors/fromBackend/AccessControlConverter.js
import { __assign, __rest } from "tslib";
import { convertWorkspaceUserGroup } from "./UserGroupsConverter";
import { convertUsersItem } from "./UsersConverter";
import { isGranularUserAccessGrantee, } from "@gooddata/sdk-model";
import isEmpty from "lodash/isEmpty";
function isGranteeUserInfo(grantee) {
    return !isEmpty(grantee) && grantee.user !== undefined;
}
export var convertGranteeEntry = function (item) {
    if (isGranteeUserInfo(item.aclEntry.grantee)) {
        return {
            type: "user",
            user: convertUsersItem(item.aclEntry.grantee.user),
        };
    }
    else {
        return {
            type: "group",
            userGroup: convertWorkspaceUserGroup(item.aclEntry.grantee.userGroup),
        };
    }
};
export var mapUserFullName = function (user) {
    if (user.fullName) {
        return user.fullName;
    }
    return user.firstName + " " + user.lastName;
};
export var convertWorkspaceUserToAvailableUserAccessGrantee = function (user) {
    var _a;
    return ({
        type: "user",
        ref: user.ref,
        name: mapUserFullName(user),
        email: user.email,
        status: (_a = user.status) !== null && _a !== void 0 ? _a : "DISABLED",
    });
};
export var convertWorkspaceUserGroupToAvailableUserGroupAccessGrantee = function (group) {
    var _a;
    return ({
        type: "group",
        ref: group.ref,
        name: (_a = group.name) !== null && _a !== void 0 ? _a : "",
    });
};
export var convertGranularAccessGranteeToAcessGrantee = function (grantee) {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    var permissions = grantee.permissions, inheritedPermissions = grantee.inheritedPermissions, rest = __rest(grantee, ["permissions", "inheritedPermissions"]);
    var type = isGranularUserAccessGrantee(grantee) ? "user" : "group";
    return __assign(__assign({}, rest), { type: type });
};
