// @gooddata/sdk-model/esm/accessControl/index.js
// (C) 2021-2023 GoodData Corporation
import isEmpty from "lodash/isEmpty";
/**
 * Tests whether the provided object is an instance of {@link IUserAccess}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isUserAccess = function (obj) {
    return !isEmpty(obj) && obj.type === "user";
};
/**
 * Tests whether the provided object is an instance of {@link IUserGroupAccess}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isUserGroupAccess = function (obj) {
    return !isEmpty(obj) && obj.type === "group";
};
/**
 * Tests whether the provided object is an instance of {@link IGranularUserAccess}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isGranularUserAccess = function (obj) {
    return !isEmpty(obj) && obj.type === "granularUser";
};
/**
 * Tests whether the provided object is an instance of {@link IGranularUserAccess}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isGranularUserGroupAccess = function (obj) {
    return !isEmpty(obj) && obj.type === "granularGroup";
};
/**
 * Tests whether the provided object is an instance of {@link IGranularUserAccess} or {@link IGranularUserGroupAccess}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isGranularAccess = function (obj) {
    return isGranularUserAccess(obj) || isGranularUserGroupAccess(obj);
};
/**
 * Tests whether the provided object is an instance of {@link IUserAccessGrantee}.
 *
 * @param obj - object to test
 * @public
 */
export var isUserAccessGrantee = function (obj) {
    return !isEmpty(obj) && obj.type === "user";
};
/**
 * Tests whether the provided object is an instance of {@link IUserGroupAccessGrantee}.
 *
 * @param obj - object to test
 * @public
 */
export var isUserGroupAccessGrantee = function (obj) {
    return !isEmpty(obj) && obj.type === "group";
};
/**
 * Tests whether the provided object is an instance of {@link IGranularUserAccessGrantee}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isGranularUserAccessGrantee = function (obj) {
    return !isEmpty(obj) && obj.type === "granularUser";
};
/**
 * Tests whether the provided object is an instance of {@link IGranularUserGroupAccessGrantee}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isGranularUserGroupAccessGrantee = function (obj) {
    return !isEmpty(obj) && obj.type === "granularGroup";
};
/**
 * Tests whether the provided object is an instance of {@link IGranularAccessGrantee}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isGranularAccessGrantee = function (obj) {
    return isGranularUserAccessGrantee(obj) || isGranularUserGroupAccessGrantee(obj);
};
/**
 * Tests whether the provided object is an instance of {@link IAvailableUserAccessGrantee}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isAvailableUserAccessGrantee = function (obj) {
    return !isEmpty(obj) && obj.type === "user";
};
/**
 * Tests whether the provided object is an instance of {@link IAvailableUserGroupAccessGrantee}.
 *
 * @param obj - object to test
 * @alpha
 */
export var isAvailableUserGroupAccessGrantee = function (obj) {
    return !isEmpty(obj) && obj.type === "group";
};
