// @gooddata/sdk-backend-tiger/esm/backend/user/settings.js
import { __assign, __awaiter, __extends, __generator } from "tslib";
import { TigerFeaturesService } from "../features";
import { DefaultUserSettings } from "../uiSettings";
import { TigerSettingsService } from "../settings";
import { unwrapSettingContent } from "../../convertors/fromBackend/SettingsConverter";
var TigerUserSettingsService = /** @class */ (function (_super) {
    __extends(TigerUserSettingsService, _super);
    function TigerUserSettingsService(authCall) {
        var _this = _super.call(this) || this;
        _this.authCall = authCall;
        return _this;
    }
    TigerUserSettingsService.prototype.getSettings = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
                        var profile, features, data, resolvedSettings;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, client.profile.getCurrent()];
                                case 1:
                                    profile = _a.sent();
                                    return [4 /*yield*/, new TigerFeaturesService(this.authCall).getFeatures(profile)];
                                case 2:
                                    features = _a.sent();
                                    return [4 /*yield*/, client.actions.resolveAllSettingsWithoutWorkspace()];
                                case 3:
                                    data = (_a.sent()).data;
                                    resolvedSettings = data.reduce(function (result, setting) {
                                        var _a;
                                        return (__assign(__assign({}, result), (_a = {}, _a[setting.id] = unwrapSettingContent(setting.content), _a)));
                                    }, {});
                                    return [2 /*return*/, __assign(__assign(__assign(__assign({}, DefaultUserSettings), features), resolvedSettings), { userId: profile.userId })];
                            }
                        });
                    }); })];
            });
        });
    };
    TigerUserSettingsService.prototype.getSettingByType = function (type) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
                        var profile;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, client.profile.getCurrent()];
                                case 1:
                                    profile = _a.sent();
                                    return [2 /*return*/, client.entities.getAllEntitiesUserSettings({
                                            userId: profile.userId,
                                            filter: "type==" + type,
                                        })];
                            }
                        });
                    }); })];
            });
        });
    };
    TigerUserSettingsService.prototype.updateSetting = function (type, id, content) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
                        var profile;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, client.profile.getCurrent()];
                                case 1:
                                    profile = _a.sent();
                                    return [2 /*return*/, client.entities.updateEntityUserSettings({
                                            id: id,
                                            userId: profile.userId,
                                            jsonApiUserSettingInDocument: {
                                                data: {
                                                    type: "userSetting",
                                                    id: id,
                                                    attributes: {
                                                        content: content,
                                                        type: type,
                                                    },
                                                },
                                            },
                                        })];
                            }
                        });
                    }); })];
            });
        });
    };
    TigerUserSettingsService.prototype.createSetting = function (type, id, content) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
                        var profile;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, client.profile.getCurrent()];
                                case 1:
                                    profile = _a.sent();
                                    return [2 /*return*/, client.entities.createEntityUserSettings({
                                            userId: profile.userId,
                                            jsonApiUserSettingInDocument: {
                                                data: {
                                                    type: "userSetting",
                                                    id: id,
                                                    attributes: {
                                                        content: content,
                                                        type: type,
                                                    },
                                                },
                                            },
                                        })];
                            }
                        });
                    }); })];
            });
        });
    };
    return TigerUserSettingsService;
}(TigerSettingsService));
export { TigerUserSettingsService };
