// @gooddata/sdk-backend-tiger/esm/backend/tigerSpecificFunctions.js
// (C) 2022-2023 GoodData Corporation
import { __assign, __awaiter, __generator, __rest } from "tslib";
import { jsonApiHeaders, JsonApiOrganizationPatchTypeEnum, MetadataUtilities, JsonApiApiTokenInTypeEnum, JsonApiWorkspaceInTypeEnum, JsonApiDataSourceInTypeEnum, OrganizationUtilities, JsonApiCspDirectiveInTypeEnum, JsonApiCustomApplicationSettingOutTypeEnum, } from "@gooddata/api-client-tiger";
import { convertApiError } from "../utils/errorHandling";
import uniq from "lodash/uniq";
import toLower from "lodash/toLower";
import { UnexpectedError } from "@gooddata/sdk-backend-spi";
import isEmpty from "lodash/isEmpty";
var getDataSourceErrorMessage = function (error) {
    if (error instanceof Error) {
        return error.message;
    }
    return String(error);
};
var dataSourceResponseAsDataSourceConnectionInfo = function (response) {
    var _a;
    var _b = response.data, id = _b.id, meta = _b.meta, attributes = _b.attributes;
    var name = attributes.name, url = attributes.url, type = attributes.type, schema = attributes.schema, username = attributes.username, parameters = attributes.parameters, decodedParameters = attributes.decodedParameters;
    return {
        id: id,
        type: type,
        name: name,
        schema: schema,
        username: username,
        url: url,
        permissions: (_a = meta === null || meta === void 0 ? void 0 : meta.permissions) !== null && _a !== void 0 ? _a : [],
        parameters: parameters,
        decodedParameters: decodedParameters,
    };
};
var dataSourceIdentifierOutWithLinksAsDataSourceConnectionInfo = function (response) {
    var _a;
    var id = response.id, meta = response.meta, attributes = response.attributes;
    var name = attributes.name, type = attributes.type, schema = attributes.schema;
    return {
        id: id,
        type: type,
        name: name,
        schema: schema,
        username: undefined,
        url: undefined,
        permissions: (_a = meta === null || meta === void 0 ? void 0 : meta.permissions) !== null && _a !== void 0 ? _a : [],
    };
};
var dataSourceIdentifierOutDocumentAsDataSourceConnectionInfo = function (response) {
    var _a;
    var _b = response.data, attributes = _b.attributes, id = _b.id, meta = _b.meta;
    var name = attributes.name, type = attributes.type, schema = attributes.schema;
    return {
        id: id,
        type: type,
        name: name,
        schema: schema,
        username: undefined,
        url: undefined,
        permissions: (_a = meta === null || meta === void 0 ? void 0 : meta.permissions) !== null && _a !== void 0 ? _a : [],
    };
};
var customAppSettingResponseAsICustomApplicationSetting = function (response) {
    var id = response.id, attributes = response.attributes;
    var applicationName = attributes.applicationName, content = attributes.content;
    return {
        id: id,
        applicationName: applicationName,
        content: content,
    };
};
export var buildTigerSpecificFunctions = function (backend, authApiCall) { return ({
    isCommunityEdition: function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var response;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.getAllEntitiesWorkspaces({ page: 0, size: 1 }, { headers: jsonApiHeaders })];
                                    case 1:
                                        response = _a.sent();
                                        // the header name are all lowercase in this object
                                        return [2 /*return*/, response.headers["gooddata-deployment"] === "aio"];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _b.sent()];
                case 2:
                    _a = _b.sent();
                    return [2 /*return*/, false];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    isOrganizationAdmin: function () { return __awaiter(void 0, void 0, void 0, function () {
        var orgPermissions, isOrganizationManage, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var organizationId, response;
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, backend.organizations().getCurrentOrganization()];
                                    case 1:
                                        organizationId = (_b.sent()).organizationId;
                                        return [4 /*yield*/, sdk.entities.getEntityOrganizations({
                                                id: organizationId,
                                                metaInclude: ["permissions"],
                                            })];
                                    case 2:
                                        response = _b.sent();
                                        return [2 /*return*/, ((_a = response.data.data.meta) === null || _a === void 0 ? void 0 : _a.permissions) || []];
                                }
                            });
                        }); })];
                case 1:
                    orgPermissions = _b.sent();
                    isOrganizationManage = function (permissions) { return permissions.includes("MANAGE"); };
                    return [2 /*return*/, isOrganizationManage(orgPermissions)];
                case 2:
                    _a = _b.sent();
                    return [2 /*return*/, false];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    organizationExpiredDate: function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var response, contractEntitlement;
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.getAllEntitiesEntitlements({})];
                                    case 1:
                                        response = _b.sent();
                                        contractEntitlement = response.data.data.find(function (item) { return item.id === "Contract"; });
                                        return [2 /*return*/, ((_a = contractEntitlement === null || contractEntitlement === void 0 ? void 0 : contractEntitlement.attributes) === null || _a === void 0 ? void 0 : _a.expiry) || ""];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _b.sent()];
                case 2:
                    _a = _b.sent();
                    return [2 /*return*/, ""];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getOrganizationAllowedOrigins: function (organizationId) { return __awaiter(void 0, void 0, void 0, function () {
        var error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            var _a, _b, _c;
                            return __generator(this, function (_d) {
                                switch (_d.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.getEntityOrganizations({ id: organizationId })];
                                    case 1:
                                        result = _d.sent();
                                        return [2 /*return*/, ((_c = (_b = (_a = result.data) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.attributes) === null || _c === void 0 ? void 0 : _c.allowedOrigins) || []];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_1 = _a.sent();
                    return [2 /*return*/, []];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getOrganizationPermissions: function (organizationId) { return __awaiter(void 0, void 0, void 0, function () {
        var error_2, toleratedCodes;
        var _a, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    _d.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            var _a, _b;
                            return __generator(this, function (_c) {
                                switch (_c.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.getEntityOrganizations({
                                            id: organizationId,
                                            metaInclude: ["permissions"],
                                        })];
                                    case 1:
                                        result = _c.sent();
                                        return [2 /*return*/, ((_b = (_a = result.data.data) === null || _a === void 0 ? void 0 : _a.meta) === null || _b === void 0 ? void 0 : _b.permissions) || []];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _d.sent()];
                case 2:
                    error_2 = _d.sent();
                    toleratedCodes = [404, 403];
                    if (toleratedCodes.includes((_a = error_2 === null || error_2 === void 0 ? void 0 : error_2.response) === null || _a === void 0 ? void 0 : _a.status) ||
                        toleratedCodes.includes((_c = (_b = error_2 === null || error_2 === void 0 ? void 0 : error_2.cause) === null || _b === void 0 ? void 0 : _b.response) === null || _c === void 0 ? void 0 : _c.status) // the error might be wrapped by an UnexpectedResponseError so check for it too
                    ) {
                        // temporary - 404 gets returned if you are not org admin
                        return [2 /*return*/, []];
                    }
                    throw convertApiError(error_2);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    updateOrganizationAllowedOrigins: function (organizationId, updatedOrigins) { return __awaiter(void 0, void 0, void 0, function () {
        var sanitizeOrigins, error_3, message;
        var _a, _b, _c, _d, _e;
        return __generator(this, function (_f) {
            switch (_f.label) {
                case 0:
                    sanitizeOrigins = function (origins) {
                        var arr = origins ? origins : [];
                        return uniq(arr.map(function (s) { return toLower(s); })).sort();
                    };
                    _f.label = 1;
                case 1:
                    _f.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            var _a, _b, _c;
                            return __generator(this, function (_d) {
                                switch (_d.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.patchEntityOrganizations({
                                            id: organizationId,
                                            jsonApiOrganizationPatchDocument: {
                                                data: {
                                                    id: organizationId,
                                                    type: JsonApiOrganizationPatchTypeEnum.ORGANIZATION,
                                                    attributes: {
                                                        allowedOrigins: updatedOrigins,
                                                    },
                                                },
                                            },
                                        })];
                                    case 1:
                                        result = _d.sent();
                                        return [2 /*return*/, sanitizeOrigins(((_c = (_b = (_a = result.data) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.attributes) === null || _c === void 0 ? void 0 : _c.allowedOrigins) || [])];
                                }
                            });
                        }); })];
                case 2: return [2 /*return*/, _f.sent()];
                case 3:
                    error_3 = _f.sent();
                    if (((_a = error_3.response) === null || _a === void 0 ? void 0 : _a.status) === 400) {
                        message = ((_c = (_b = error_3 === null || error_3 === void 0 ? void 0 : error_3.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.detail) ? (_e = (_d = error_3 === null || error_3 === void 0 ? void 0 : error_3.response) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.detail : "Server error";
                        throw new UnexpectedError(message, error_3);
                    }
                    throw convertApiError(error_3);
                case 4: return [2 /*return*/];
            }
        });
    }); },
    getDeploymentVersion: function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var profile;
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, sdk.axios.get("/api/v1/profile")];
                                    case 1:
                                        profile = _b.sent();
                                        return [2 /*return*/, (_a = profile === null || profile === void 0 ? void 0 : profile.headers) === null || _a === void 0 ? void 0 : _a["gooddata-deployment"]];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _b.sent()];
                case 2:
                    _a = _b.sent();
                    return [2 /*return*/, undefined];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getAllApiTokens: function (userId) { return __awaiter(void 0, void 0, void 0, function () {
        var mapTokens, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    mapTokens = function (tokens) {
                        return tokens.data.map(function (item) {
                            var result = {
                                id: item.id,
                            };
                            return result;
                        });
                    };
                    _b.label = 1;
                case 1:
                    _b.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, MetadataUtilities.getAllPagesOf(sdk, sdk.entities.getAllEntitiesApiTokens, {
                                            userId: userId,
                                        })
                                            .then(MetadataUtilities.mergeEntitiesResults)
                                            .then(mapTokens)];
                                    case 1: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                case 2: return [2 /*return*/, _b.sent()];
                case 3:
                    _a = _b.sent();
                    return [2 /*return*/, []];
                case 4: return [2 /*return*/];
            }
        });
    }); },
    generateApiToken: function (userId, tokenId) { return __awaiter(void 0, void 0, void 0, function () {
        var error_4;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var apiTokenDocument, result;
                            var _a, _b, _c;
                            return __generator(this, function (_d) {
                                switch (_d.label) {
                                    case 0:
                                        apiTokenDocument = {
                                            data: {
                                                id: tokenId,
                                                type: JsonApiApiTokenInTypeEnum.API_TOKEN,
                                            },
                                        };
                                        return [4 /*yield*/, sdk.entities.createEntityApiTokens({
                                                userId: userId,
                                                jsonApiApiTokenInDocument: apiTokenDocument,
                                            })];
                                    case 1:
                                        result = _d.sent();
                                        return [2 /*return*/, {
                                                id: (_a = result.data.data) === null || _a === void 0 ? void 0 : _a.id,
                                                bearerToken: (_c = (_b = result.data.data) === null || _b === void 0 ? void 0 : _b.attributes) === null || _c === void 0 ? void 0 : _c.bearerToken,
                                            }];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_4 = _a.sent();
                    throw convertApiError(error_4);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    deleteApiToken: function (userId, tokenId) { return __awaiter(void 0, void 0, void 0, function () {
        var error_5;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.deleteEntityApiTokens({ userId: userId, id: tokenId })];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    error_5 = _a.sent();
                    throw convertApiError(error_5);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    someDataSourcesExists: function (filter) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        var requestParams;
                        return __generator(this, function (_a) {
                            requestParams = filter ? { filter: filter } : {};
                            return [2 /*return*/, sdk.entities
                                    .getAllEntitiesDataSources(requestParams)
                                    .then(function (axiosResponse) { var _a; return ((_a = axiosResponse.data.data) === null || _a === void 0 ? void 0 : _a.length) > 0; })
                                    .catch(function () {
                                    return false;
                                })];
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    generateLogicalModel: function (dataSourceId, generateLdmRequest) { return __awaiter(void 0, void 0, void 0, function () {
        var error_6;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                return [2 /*return*/, sdk.actions
                                        .generateLogicalModel({
                                        dataSourceId: dataSourceId,
                                        generateLdmRequest: generateLdmRequest,
                                    })
                                        .then(function (axiosResponse) { return axiosResponse.data; })];
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_6 = _a.sent();
                    throw convertApiError(error_6);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    scanDataSource: function (dataSourceId, scanRequest) { return __awaiter(void 0, void 0, void 0, function () {
        var error_7;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.scanModel
                                            .scanDataSource({
                                            dataSourceId: dataSourceId,
                                            scanRequest: scanRequest,
                                        })
                                            .then(function (res) {
                                            return res === null || res === void 0 ? void 0 : res.data;
                                        })];
                                    case 1: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_7 = _a.sent();
                    throw convertApiError(error_7);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    publishPdm: function (dataSourceId, declarativePdm) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, sdk.declarativeLayout
                                        .setPdmLayout({
                                        dataSourceId: dataSourceId,
                                        declarativePdm: declarativePdm,
                                    })
                                        .then(function () {
                                        return {
                                            status: "success",
                                        };
                                    }, function (error) {
                                        return Promise.reject(error);
                                    })];
                                case 1: return [2 /*return*/, _a.sent()];
                            }
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    createDemoWorkspace: function (sampleWorkspace) { return __awaiter(void 0, void 0, void 0, function () {
        var error_8;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.createEntityWorkspaces({
                                            jsonApiWorkspaceInDocument: sampleWorkspace,
                                        })];
                                    case 1:
                                        result = _a.sent();
                                        return [2 /*return*/, result.data.data.id];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_8 = _a.sent();
                    throw convertApiError(error_8);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    createDemoDataSource: function (sampleDataSource) { return __awaiter(void 0, void 0, void 0, function () {
        var error_9;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.createEntityDataSources({
                                            jsonApiDataSourceInDocument: sampleDataSource,
                                        })];
                                    case 1:
                                        result = _a.sent();
                                        return [2 /*return*/, result.data.data.id];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_9 = _a.sent();
                    throw convertApiError(error_9);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    setPdmLayout: function (requestParameters) { return __awaiter(void 0, void 0, void 0, function () {
        var error_10;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.declarativeLayout.setPdmLayout(requestParameters)];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_10 = _a.sent();
                    throw convertApiError(error_10);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    createWorkspace: function (id, name) { return __awaiter(void 0, void 0, void 0, function () {
        var error_11;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.createEntityWorkspaces({
                                            jsonApiWorkspaceInDocument: {
                                                data: {
                                                    attributes: {
                                                        name: name,
                                                    },
                                                    id: id,
                                                    type: JsonApiWorkspaceInTypeEnum.WORKSPACE,
                                                },
                                            },
                                        })];
                                    case 1:
                                        result = _a.sent();
                                        return [2 /*return*/, result.data.data.id];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_11 = _a.sent();
                    throw convertApiError(error_11);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    updateWorkspaceTitle: function (id, name) { return __awaiter(void 0, void 0, void 0, function () {
        var error_12;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.updateEntityWorkspaces({
                                            id: id,
                                            jsonApiWorkspaceInDocument: {
                                                data: {
                                                    attributes: {
                                                        name: name,
                                                    },
                                                    id: id,
                                                    type: JsonApiWorkspaceInTypeEnum.WORKSPACE,
                                                },
                                            },
                                        })];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_12 = _a.sent();
                    throw convertApiError(error_12);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    deleteWorkspace: function (id) { return __awaiter(void 0, void 0, void 0, function () {
        var error_13;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.deleteEntityWorkspaces({ id: id })];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_13 = _a.sent();
                    throw convertApiError(error_13);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    canDeleteWorkspace: function (id) { return __awaiter(void 0, void 0, void 0, function () {
        var e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var childWorkspaces;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.getAllEntitiesWorkspaces({
                                            include: ["workspaces"],
                                            filter: "parent.id==" + id,
                                        })];
                                    case 1:
                                        childWorkspaces = (_a.sent()).data.data;
                                        return [2 /*return*/, isEmpty(childWorkspaces)];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    e_1 = _a.sent();
                    return [2 /*return*/, true];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getWorkspaceLogicalModel: function (workspaceId, includeParents) {
        if (includeParents === void 0) { includeParents = false; }
        return __awaiter(void 0, void 0, void 0, function () {
            var error_14;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                                var result;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, sdk.declarativeLayout.getLogicalModel({ workspaceId: workspaceId, includeParents: includeParents })];
                                        case 1:
                                            result = _a.sent();
                                            return [2 /*return*/, result.data];
                                    }
                                });
                            }); })];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2:
                        error_14 = _a.sent();
                        throw convertApiError(error_14);
                    case 3: return [2 /*return*/];
                }
            });
        });
    },
    putWorkspaceLayout: function (requestParameters) { return __awaiter(void 0, void 0, void 0, function () {
        var error_15;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.declarativeLayout.putWorkspaceLayout(requestParameters)];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_15 = _a.sent();
                    throw convertApiError(error_15);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getEntitlements: function () { return __awaiter(void 0, void 0, void 0, function () {
        var error_16;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.getAllEntitiesEntitlements({})];
                                    case 1:
                                        result = _a.sent();
                                        return [2 /*return*/, result.data.data.map(function (entitlement) {
                                                var _a, _b;
                                                return ({
                                                    id: entitlement.id,
                                                    value: (_a = entitlement.attributes) === null || _a === void 0 ? void 0 : _a.value,
                                                    expiry: (_b = entitlement.attributes) === null || _b === void 0 ? void 0 : _b.expiry,
                                                });
                                            })];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_16 = _a.sent();
                    throw convertApiError(error_16);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getDataSourceById: function (id) { return __awaiter(void 0, void 0, void 0, function () {
        var error_17;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                return [2 /*return*/, sdk.entities
                                        .getEntityDataSources({
                                        id: id,
                                    })
                                        .then(function (axiosResponse) { return ({
                                        data: dataSourceResponseAsDataSourceConnectionInfo(axiosResponse.data),
                                    }); })];
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_17 = _a.sent();
                    return [2 /*return*/, { errorMessage: getDataSourceErrorMessage(error_17) }];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getDataSourceIdentifierById: function (id) { return __awaiter(void 0, void 0, void 0, function () {
        var error_18;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                return [2 /*return*/, sdk.entities
                                        .getEntityDataSourceIdentifiers({
                                        id: id,
                                    })
                                        .then(function (axiosResponse) { return ({
                                        data: dataSourceIdentifierOutDocumentAsDataSourceConnectionInfo(axiosResponse.data),
                                    }); })];
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_18 = _a.sent();
                    return [2 /*return*/, { errorMessage: getDataSourceErrorMessage(error_18) }];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    createDataSource: function (requestData) { return __awaiter(void 0, void 0, void 0, function () {
        var id, name, password, schema, token, type, url, username, parameters, error_19;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    id = requestData.id, name = requestData.name, password = requestData.password, schema = requestData.schema, token = requestData.token, type = requestData.type, url = requestData.url, username = requestData.username, parameters = requestData.parameters;
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                return [2 /*return*/, sdk.entities
                                        .createEntityDataSources({
                                        jsonApiDataSourceInDocument: {
                                            data: {
                                                attributes: {
                                                    name: name,
                                                    password: password,
                                                    schema: schema,
                                                    token: token,
                                                    type: type,
                                                    url: url,
                                                    username: username,
                                                    parameters: parameters,
                                                },
                                                id: id,
                                                type: JsonApiDataSourceInTypeEnum.DATA_SOURCE,
                                            },
                                        },
                                    })
                                        .then(function (axiosResponse) { return ({
                                        data: dataSourceResponseAsDataSourceConnectionInfo(axiosResponse.data),
                                    }); })];
                            });
                        }); })];
                case 2: return [2 /*return*/, _a.sent()];
                case 3:
                    error_19 = _a.sent();
                    return [2 /*return*/, { errorMessage: getDataSourceErrorMessage(error_19) }];
                case 4: return [2 /*return*/];
            }
        });
    }); },
    updateDataSource: function (id, requestData) { return __awaiter(void 0, void 0, void 0, function () {
        var requestDataId, name, password, schema, token, type, url, username, parameters, error_20;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    requestDataId = requestData.id, name = requestData.name, password = requestData.password, schema = requestData.schema, token = requestData.token, type = requestData.type, url = requestData.url, username = requestData.username, parameters = requestData.parameters;
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                return [2 /*return*/, sdk.entities
                                        .updateEntityDataSources({
                                        id: id,
                                        jsonApiDataSourceInDocument: {
                                            data: {
                                                attributes: {
                                                    name: name,
                                                    password: password,
                                                    schema: schema,
                                                    token: token,
                                                    type: type,
                                                    url: url,
                                                    username: username,
                                                    parameters: parameters,
                                                },
                                                id: requestDataId,
                                                type: JsonApiDataSourceInTypeEnum.DATA_SOURCE,
                                            },
                                        },
                                    })
                                        .then(function (axiosResponse) { return ({
                                        data: dataSourceResponseAsDataSourceConnectionInfo(axiosResponse.data),
                                    }); })];
                            });
                        }); })];
                case 2: return [2 /*return*/, _a.sent()];
                case 3:
                    error_20 = _a.sent();
                    return [2 /*return*/, { errorMessage: getDataSourceErrorMessage(error_20) }];
                case 4: return [2 /*return*/];
            }
        });
    }); },
    patchDataSource: function (id, requestData) { return __awaiter(void 0, void 0, void 0, function () {
        var requestDataId, name, password, schema, token, type, url, username, parameters, error_21;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    requestDataId = requestData.id, name = requestData.name, password = requestData.password, schema = requestData.schema, token = requestData.token, type = requestData.type, url = requestData.url, username = requestData.username, parameters = requestData.parameters;
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                return [2 /*return*/, sdk.entities
                                        .patchEntityDataSources({
                                        id: id,
                                        jsonApiDataSourcePatchDocument: {
                                            data: {
                                                attributes: {
                                                    name: name,
                                                    password: password,
                                                    schema: schema,
                                                    token: token,
                                                    type: type,
                                                    url: url,
                                                    username: username,
                                                    parameters: parameters,
                                                },
                                                id: requestDataId,
                                                type: JsonApiDataSourceInTypeEnum.DATA_SOURCE,
                                            },
                                        },
                                    })
                                        .then(function (axiosResponse) { return ({
                                        data: dataSourceResponseAsDataSourceConnectionInfo(axiosResponse.data),
                                    }); })];
                            });
                        }); })];
                case 2: return [2 /*return*/, _a.sent()];
                case 3:
                    error_21 = _a.sent();
                    return [2 /*return*/, { errorMessage: getDataSourceErrorMessage(error_21) }];
                case 4: return [2 /*return*/];
            }
        });
    }); },
    deleteDataSource: function (id) { return __awaiter(void 0, void 0, void 0, function () {
        var error_22;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.deleteEntityDataSources({
                                            id: id,
                                        })];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/, { successful: true }];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_22 = _a.sent();
                    return [2 /*return*/, { errorMessage: getDataSourceErrorMessage(error_22) }];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    testDataSourceConnection: function (connectionData, id) { return __awaiter(void 0, void 0, void 0, function () {
        var error_23;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var promise;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        promise = id
                                            ? sdk.scanModel.testDataSource({
                                                dataSourceId: id,
                                                testRequest: connectionData,
                                            })
                                            : sdk.scanModel.testDataSourceDefinition({
                                                testDefinitionRequest: connectionData,
                                            });
                                        return [4 /*yield*/, promise.then(function (axiosResponse) { return axiosResponse.data; })];
                                    case 1: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_23 = _a.sent();
                    return [2 /*return*/, {
                            successful: false,
                            error: getDataSourceErrorMessage(error_23),
                        }];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getDataSourceSchemata: function (dataSourceId) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, sdk.scanModel.getDataSourceSchemata({ dataSourceId: dataSourceId }).then(function (res) {
                                        return res === null || res === void 0 ? void 0 : res.data.schemaNames;
                                    })];
                                case 1: return [2 /*return*/, _a.sent()];
                            }
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    getPdm: function (dataSourceId) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, sdk.declarativeLayout
                                        .getPdmLayout({
                                        dataSourceId: dataSourceId,
                                    })
                                        .then(function (res) {
                                        return res === null || res === void 0 ? void 0 : res.data;
                                    })];
                                case 1: return [2 /*return*/, _a.sent()];
                            }
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    getAllDataSources: function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            return [2 /*return*/, OrganizationUtilities.getAllPagesOf(sdk, sdk.entities.getAllEntitiesDataSourceIdentifiers, {
                                    sort: ["name"],
                                    metaInclude: ["permissions"],
                                })
                                    .then(OrganizationUtilities.mergeEntitiesResults)
                                    .then(function (res) {
                                    return res.data.map(dataSourceIdentifierOutWithLinksAsDataSourceConnectionInfo);
                                })];
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    publishLogicalModel: function (workspaceId, declarativeModel) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, sdk.declarativeLayout.setLogicalModel({
                                        workspaceId: workspaceId,
                                        declarativeModel: declarativeModel,
                                    })];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    getDependentEntitiesGraph: function (workspaceId) { return __awaiter(void 0, void 0, void 0, function () {
        var error_24;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.actions
                                            .getDependentEntitiesGraph({
                                            workspaceId: workspaceId,
                                        })
                                            .then(function (res) {
                                            return res === null || res === void 0 ? void 0 : res.data;
                                        })];
                                    case 1: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_24 = _a.sent();
                    throw convertApiError(error_24);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getDependentEntitiesGraphFromEntryPoints: function (workspaceId, dependentEntitiesGraphRequest) { return __awaiter(void 0, void 0, void 0, function () {
        var error_25;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.actions
                                            .getDependentEntitiesGraphFromEntryPoints({
                                            workspaceId: workspaceId,
                                            dependentEntitiesRequest: dependentEntitiesGraphRequest,
                                        })
                                            .then(function (res) {
                                            return res === null || res === void 0 ? void 0 : res.data;
                                        })];
                                    case 1: return [2 /*return*/, _a.sent()];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_25 = _a.sent();
                    throw convertApiError(error_25);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    resolveAllEntitlements: function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, sdk.actions.resolveAllEntitlements().then(function (res) { return res.data; })];
                }); }); })];
        });
    }); },
    getAllPlatformUsage: function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, sdk.actions.allPlatformUsage().then(function (res) { return res.data; })];
                }); }); })];
        });
    }); },
    inviteUser: function (requestParameters, options) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        return [2 /*return*/, sdk.authActions.processInvitation(requestParameters, options).then(function (res) {
                                if (res.status == 204) {
                                    return {
                                        successful: true,
                                    };
                                }
                                else {
                                    return {
                                        successful: false,
                                        errorMessage: res === null || res === void 0 ? void 0 : res.data,
                                    };
                                }
                            })];
                    });
                }); })];
        });
    }); },
    getWorkspaceDataFiltersLayout: function () { return __awaiter(void 0, void 0, void 0, function () {
        var error_26;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.declarativeLayout.getWorkspaceDataFiltersLayout()];
                                    case 1:
                                        result = _a.sent();
                                        return [2 /*return*/, result.data];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_26 = _a.sent();
                    throw convertApiError(error_26);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    setWorkspaceDataFiltersLayout: function (workspaceDataFiltersLayout) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, sdk.declarativeLayout.setWorkspaceDataFiltersLayout({
                                        declarativeWorkspaceDataFilters: workspaceDataFiltersLayout,
                                    })];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    getAllCSPDirectives: function () { return __awaiter(void 0, void 0, void 0, function () {
        var error_27;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.getAllEntitiesCspDirectives({})];
                                    case 1:
                                        result = _b.sent();
                                        return [2 /*return*/, ((_a = result.data) === null || _a === void 0 ? void 0 : _a.data) || []];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_27 = _a.sent();
                    return [2 /*return*/, []];
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getCSPDirective: function (directiveId) { return __awaiter(void 0, void 0, void 0, function () {
        var error_28;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result;
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.getEntityCspDirectives({ id: directiveId })];
                                    case 1:
                                        result = _b.sent();
                                        return [2 /*return*/, (_a = result.data) === null || _a === void 0 ? void 0 : _a.data];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_28 = _a.sent();
                    throw convertApiError(error_28);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    createCSPDirective: function (requestData) { return __awaiter(void 0, void 0, void 0, function () {
        var error_29, message;
        var _a, _b, _c, _d, _e;
        return __generator(this, function (_f) {
            switch (_f.label) {
                case 0:
                    _f.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var jsonApiCspDirectiveInDocument, result;
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        jsonApiCspDirectiveInDocument = {
                                            data: {
                                                id: requestData.id,
                                                type: JsonApiCspDirectiveInTypeEnum.CSP_DIRECTIVE,
                                                attributes: requestData.attributes,
                                            },
                                        };
                                        return [4 /*yield*/, sdk.entities.createEntityCspDirectives({
                                                jsonApiCspDirectiveInDocument: jsonApiCspDirectiveInDocument,
                                            })];
                                    case 1:
                                        result = _b.sent();
                                        return [2 /*return*/, (_a = result.data) === null || _a === void 0 ? void 0 : _a.data];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _f.sent()];
                case 2:
                    error_29 = _f.sent();
                    if (((_a = error_29.response) === null || _a === void 0 ? void 0 : _a.status) === 400) {
                        message = ((_c = (_b = error_29 === null || error_29 === void 0 ? void 0 : error_29.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.detail) ? (_e = (_d = error_29 === null || error_29 === void 0 ? void 0 : error_29.response) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.detail : "Server error";
                        throw new UnexpectedError(message, error_29);
                    }
                    throw convertApiError(error_29);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    updateCSPDirective: function (directiveId, requestData) { return __awaiter(void 0, void 0, void 0, function () {
        var error_30, message;
        var _a, _b, _c, _d, _e;
        return __generator(this, function (_f) {
            switch (_f.label) {
                case 0:
                    _f.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var jsonApiCspDirectiveInDocument, result;
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        jsonApiCspDirectiveInDocument = {
                                            data: {
                                                id: requestData.id,
                                                type: JsonApiCspDirectiveInTypeEnum.CSP_DIRECTIVE,
                                                attributes: requestData.attributes,
                                            },
                                        };
                                        return [4 /*yield*/, sdk.entities.updateEntityCspDirectives({
                                                id: directiveId,
                                                jsonApiCspDirectiveInDocument: jsonApiCspDirectiveInDocument,
                                            })];
                                    case 1:
                                        result = _b.sent();
                                        return [2 /*return*/, (_a = result.data) === null || _a === void 0 ? void 0 : _a.data];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _f.sent()];
                case 2:
                    error_30 = _f.sent();
                    if (((_a = error_30.response) === null || _a === void 0 ? void 0 : _a.status) === 400) {
                        message = ((_c = (_b = error_30 === null || error_30 === void 0 ? void 0 : error_30.response) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.detail) ? (_e = (_d = error_30 === null || error_30 === void 0 ? void 0 : error_30.response) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.detail : "Server error";
                        throw new UnexpectedError(message, error_30);
                    }
                    throw convertApiError(error_30);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    deleteCSPDirective: function (directiveId) { return __awaiter(void 0, void 0, void 0, function () {
        var error_31;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.deleteEntityCspDirectives({ id: directiveId })];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 1:
                    _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    error_31 = _a.sent();
                    throw convertApiError(error_31);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getWorkspaceCustomAppSettings: function (workspaceId, applicationName) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        var result, responseData;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, sdk.entities.getAllEntitiesCustomApplicationSettings({
                                        workspaceId: workspaceId,
                                    })];
                                case 1:
                                    result = _a.sent();
                                    responseData = result.data;
                                    if (applicationName) {
                                        responseData.data = responseData.data.filter(function (setting) { return setting.attributes.applicationName === applicationName; });
                                    }
                                    return [2 /*return*/, responseData.data.map(function (setting) {
                                            return customAppSettingResponseAsICustomApplicationSetting(setting);
                                        })];
                            }
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    createWorkspaceCustomAppSetting: function (workspaceId, id, applicationName, content) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        var result;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, sdk.entities.createEntityCustomApplicationSettings({
                                        workspaceId: workspaceId,
                                        jsonApiCustomApplicationSettingInDocument: {
                                            data: {
                                                id: id,
                                                type: JsonApiCustomApplicationSettingOutTypeEnum.CUSTOM_APPLICATION_SETTING,
                                                attributes: {
                                                    applicationName: applicationName,
                                                    content: content,
                                                },
                                            },
                                        },
                                    })];
                                case 1:
                                    result = _a.sent();
                                    return [2 /*return*/, customAppSettingResponseAsICustomApplicationSetting(result.data.data)];
                            }
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    deleteWorkspaceCustomAppSetting: function (workspaceId, settingId) { return __awaiter(void 0, void 0, void 0, function () {
        var error_32;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.deleteEntityCustomApplicationSettings({
                                            objectId: settingId,
                                            workspaceId: workspaceId,
                                        })];
                                    case 1:
                                        _a.sent();
                                        return [2 /*return*/];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_32 = _a.sent();
                    throw convertApiError(error_32);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    getEntityUser: function (id) { return __awaiter(void 0, void 0, void 0, function () {
        var error_33;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            var result, _a, firstname, lastname, userInfo;
                            var _b;
                            return __generator(this, function (_c) {
                                switch (_c.label) {
                                    case 0: return [4 /*yield*/, sdk.entities.getEntityUsers({
                                            id: id,
                                        })];
                                    case 1:
                                        result = _c.sent();
                                        _a = ((_b = result.data) === null || _b === void 0 ? void 0 : _b.data.attributes) || {}, firstname = _a.firstname, lastname = _a.lastname, userInfo = __rest(_a, ["firstname", "lastname"]);
                                        return [2 /*return*/, __assign(__assign({}, userInfo), { firstName: firstname, lastName: lastname })];
                                }
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_33 = _a.sent();
                    throw convertApiError(error_33);
                case 3: return [2 /*return*/];
            }
        });
    }); },
    registerUploadNotification: function (dataSourceId) { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, sdk.actions.registerUploadNotification({
                                        dataSourceId: dataSourceId,
                                    })];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
                case 1: return [2 /*return*/, _a.sent()];
            }
        });
    }); },
    scanSql: function (dataSourceId, sql) { return __awaiter(void 0, void 0, void 0, function () {
        var error_34;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, authApiCall(function (sdk) { return __awaiter(void 0, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                return [2 /*return*/, sdk.scanModel.scanSql({ dataSourceId: dataSourceId, scanSqlRequest: { sql: sql } }).then(function (response) {
                                        return response.data;
                                    })];
                            });
                        }); })];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    error_34 = _a.sent();
                    throw convertApiError(error_34);
                case 3: return [2 /*return*/];
            }
        });
    }); },
}); };
