// @gooddata/sdk-backend-tiger/esm/backend/uiFeatures.js
// (C) 2020-2023 GoodData Corporation
/**
 * This is list of feature flags managed on Panther by FeatureHub
 * and keeping their default values for non-managed env
 */
export var TigerFeaturesNames;
(function (TigerFeaturesNames) {
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableSortingByTotalGroup"] = "enableSortingByTotalGroup";
    //string, possible values: disabled, enabledCheckedByDefault, enabledUncheckedByDefault
    TigerFeaturesNames["ADMeasureValueFilterNullAsZeroOption"] = "ADMeasureValueFilterNullAsZeroOption";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableMultipleDates"] = "enableMultipleDates";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableKPIDashboardDeleteFilterButton"] = "enableKPIDashboardDeleteFilterButton";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["DashboardEditModeDevRollout"] = "dashboardEditModeDevRollout";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableMetricSqlAndDataExplain"] = "enableMetricSqlAndDataExplain";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableLongitudeAndLatitudeLabels"] = "enableLongitudeAndLatitudeLabels";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableDescriptions"] = "enableDescriptions";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableAnalyticalDashboardPermissions"] = "enableAnalyticalDashboardPermissions";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnablePushpinGeoChart"] = "enablePushpinGeoChart";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableKPIAttributeFilterRenaming"] = "enableKPIAttributeFilterRenaming";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableKPIDashboardExportPDF"] = "enableKPIDashboardExportPDF";
    //boolean + possible values: enabled, disabled
    TigerFeaturesNames["EnableSqlDatasets"] = "enableSqlDatasets";
})(TigerFeaturesNames || (TigerFeaturesNames = {}));
export var DefaultFeatureFlags = {
    enableSortingByTotalGroup: false,
    ADMeasureValueFilterNullAsZeroOption: "EnabledUncheckedByDefault",
    enableMultipleDates: true,
    enableKPIDashboardDeleteFilterButton: false,
    dashboardEditModeDevRollout: true,
    enableMetricSqlAndDataExplain: false,
    enableLongitudeAndLatitudeLabels: true,
    enableDescriptions: true,
    enableAnalyticalDashboardPermissions: true,
    enablePushpinGeoChart: true,
    enableKPIAttributeFilterRenaming: false,
    enableKPIDashboardExportPDF: true,
    enableSqlDatasets: false,
};
export var FeatureFlagsValues = {
    enableSortingByTotalGroup: [true, false],
    ADMeasureValueFilterNullAsZeroOption: [
        "Disabled",
        "EnabledCheckedByDefault",
        "EnabledUncheckedByDefault",
    ],
    enableMultipleDates: [true, false],
    enableKPIDashboardDeleteFilterButton: [true, false],
    dashboardEditModeDevRollout: [true, false],
    enableMetricSqlAndDataExplain: [true, false],
    enableLongitudeAndLatitudeLabels: [true, false],
    enableDescriptions: [true, false],
    enableAnalyticalDashboardPermissions: [true, false],
    enablePushpinGeoChart: [true, false],
    enableKPIAttributeFilterRenaming: [true, false],
    enableKPIDashboardExportPDF: [true, false],
    enableSqlDatasets: [true, false],
};
