// @gooddata/sdk-backend-tiger/esm/backend/workspace/settings/index.js
import { __assign, __awaiter, __extends, __generator } from "tslib";
import { TigerFeaturesService, pickContext } from "../../features";
import { DefaultUiSettings, DefaultUserSettings } from "../../uiSettings";
import { unwrapSettingContent } from "../../../convertors/fromBackend/SettingsConverter";
import { TigerSettingsService } from "../../settings";
var TigerWorkspaceSettings = /** @class */ (function (_super) {
    __extends(TigerWorkspaceSettings, _super);
    function TigerWorkspaceSettings(authCall, workspace) {
        var _this = _super.call(this) || this;
        _this.authCall = authCall;
        _this.workspace = workspace;
        return _this;
    }
    TigerWorkspaceSettings.prototype.getSettings = function () {
        var _this = this;
        return this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
            var config;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, client.entities.getEntityWorkspaces({
                            id: this.workspace,
                            metaInclude: ["config"],
                        })];
                    case 1:
                        config = (_a.sent()).data.data.meta;
                        return [2 /*return*/, __assign(__assign({ workspace: this.workspace }, DefaultUiSettings), config === null || config === void 0 ? void 0 : config.config)];
                }
            });
        }); });
    };
    TigerWorkspaceSettings.prototype.getSettingsForCurrentUser = function () {
        return getSettingsForCurrentUser(this.authCall, this.workspace);
    };
    TigerWorkspaceSettings.prototype.getSettingByType = function (type) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) {
                        return client.entities.getAllEntitiesWorkspaceSettings({
                            workspaceId: _this.workspace,
                            origin: "NATIVE",
                            filter: "type==" + type,
                        });
                    })];
            });
        });
    };
    TigerWorkspaceSettings.prototype.updateSetting = function (type, id, content) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            return [2 /*return*/, client.entities.updateEntityWorkspaceSettings({
                                    workspaceId: this.workspace,
                                    objectId: id,
                                    jsonApiWorkspaceSettingInDocument: {
                                        data: {
                                            type: "workspaceSetting",
                                            id: id,
                                            attributes: {
                                                content: content,
                                                type: type,
                                            },
                                        },
                                    },
                                })];
                        });
                    }); })];
            });
        });
    };
    TigerWorkspaceSettings.prototype.createSetting = function (type, id, content) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, this.authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            return [2 /*return*/, client.entities.createEntityWorkspaceSettings({
                                    workspaceId: this.workspace,
                                    jsonApiWorkspaceSettingInDocument: {
                                        data: {
                                            type: "workspaceSetting",
                                            id: id,
                                            attributes: {
                                                content: content,
                                                type: type,
                                            },
                                        },
                                    },
                                })];
                        });
                    }); })];
            });
        });
    };
    return TigerWorkspaceSettings;
}(TigerSettingsService));
export { TigerWorkspaceSettings };
/**
 * @internal
 */
function resolveSettings(authCall, workspace) {
    return __awaiter(this, void 0, void 0, function () {
        var data;
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            return [2 /*return*/, client.actions.workspaceResolveAllSettings({
                                    workspaceId: workspace,
                                })];
                        });
                    }); })];
                case 1:
                    data = (_a.sent()).data;
                    return [2 /*return*/, data.reduce(function (result, setting) {
                            var _a;
                            return __assign(__assign({}, result), (_a = {}, _a[setting.id] = unwrapSettingContent(setting.content), _a));
                        }, {})];
            }
        });
    });
}
/**
 * Expose this wrapper to other SPI implementations
 *
 * @internal
 */
export function getSettingsForCurrentUser(authCall, workspace) {
    var _this = this;
    return authCall(function (client) { return __awaiter(_this, void 0, void 0, function () {
        var _a, config, attributes, resolvedSettings, profile, features;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, client.entities.getEntityWorkspaces({
                        id: workspace,
                        metaInclude: ["config"],
                    })];
                case 1:
                    _a = (_b.sent()).data.data, config = _a.meta, attributes = _a.attributes;
                    return [4 /*yield*/, resolveSettings(authCall, workspace)];
                case 2:
                    resolvedSettings = _b.sent();
                    return [4 /*yield*/, client.profile.getCurrent()];
                case 3:
                    profile = _b.sent();
                    return [4 /*yield*/, new TigerFeaturesService(authCall).getFeatures(profile, pickContext(attributes))];
                case 4:
                    features = _b.sent();
                    return [2 /*return*/, __assign(__assign(__assign(__assign(__assign({}, DefaultUserSettings), { userId: profile.userId, workspace: workspace }), config === null || config === void 0 ? void 0 : config.config), features), resolvedSettings)];
            }
        });
    }); });
}
