// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/bubbleChart/BubbleChartDescriptor.js
import { __extends } from "tslib";
import { BucketNames } from "@gooddata/sdk-ui";
import { PluggableBubbleChart } from "./PluggableBubbleChart";
import { BigChartDescriptor } from "../BigChartDescriptor";
import { executionConfigInsightConversion, filtersInsightConversion, getInsightToPropsConverter, getReactEmbeddingCodeGenerator, localeInsightConversion, singleAttributeBucketConversion, singleMeasureBucketConversion, sortsInsightConversion, } from "../../../utils/embeddingCodeGenerator";
import { chartAdditionalFactories, chartConfigInsightConversion } from "../chartCodeGenUtils";
var BubbleChartDescriptor = /** @class */ (function (_super) {
    __extends(BubbleChartDescriptor, _super);
    function BubbleChartDescriptor() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.getEmbeddingCode = getReactEmbeddingCodeGenerator({
            component: {
                importType: "named",
                name: "BubbleChart",
                package: "@gooddata/sdk-ui-charts",
            },
            insightToProps: getInsightToPropsConverter({
                xAxisMeasure: singleMeasureBucketConversion("xAxisMeasure", BucketNames.MEASURES),
                yAxisMeasure: singleMeasureBucketConversion("yAxisMeasure", BucketNames.SECONDARY_MEASURES),
                size: singleMeasureBucketConversion("size", BucketNames.TERTIARY_MEASURES),
                viewBy: singleAttributeBucketConversion("viewBy", BucketNames.VIEW),
                filters: filtersInsightConversion("filters"),
                sortBy: sortsInsightConversion("sortBy"),
                config: chartConfigInsightConversion("config"),
                locale: localeInsightConversion("locale"),
                execConfig: executionConfigInsightConversion("execConfig"),
            }),
            additionalFactories: chartAdditionalFactories(),
        });
        return _this;
    }
    BubbleChartDescriptor.prototype.getFactory = function () {
        return function (params) { return new PluggableBubbleChart(params); };
    };
    BubbleChartDescriptor.prototype.getMeta = function () {
        return {
            documentationUrl: "https://sdk.gooddata.com/gooddata-ui/docs/bubble_chart_component.html",
            supportsExport: true,
        };
    };
    return BubbleChartDescriptor;
}(BigChartDescriptor));
export { BubbleChartDescriptor };
