// @gooddata/sdk-ui-kit/esm/Overlay/OverlayController.js
// (C) 2022 GoodData Corporation
/**
 * When stacking the overlays, next overlay is going to have an z-index higher by 1.
 */
var Z_INDEX_STEP = 1;
var DEFAULT_INIT_Z_INDEX = 5001;
/**
 * Overlay stacking controller.
 *
 * @internal
 */
var OverlayController = /** @class */ (function () {
    function OverlayController(initialZIndex) {
        if (initialZIndex === void 0) { initialZIndex = DEFAULT_INIT_Z_INDEX; }
        this.initialZIndex = DEFAULT_INIT_Z_INDEX;
        this.overlays = new Map();
        this.initialZIndex = initialZIndex;
    }
    OverlayController.getInstance = function (initialZIndex) {
        if (initialZIndex === void 0) { initialZIndex = DEFAULT_INIT_Z_INDEX; }
        return new OverlayController(initialZIndex);
    };
    /**
     * Adds the overlay record to the map.
     *
     * @param uuid - given overlay uuid.
     */
    OverlayController.prototype.addOverlay = function (uuid) {
        var maxIndex = this.getMaxZIndex();
        this.overlays.set(uuid, maxIndex + Z_INDEX_STEP);
    };
    /**
     * Get maximum z-Index from current opened overlays
     * @returns
     */
    OverlayController.prototype.getMaxZIndex = function () {
        var maxIndex = 0;
        this.overlays.forEach(function (value) {
            maxIndex = Math.max(maxIndex, value);
        });
        return maxIndex || this.initialZIndex;
    };
    /**
     * Getter for z-index of the given overlay.
     *
     * @remarks
     * If the entry in the overlays map is not available, the initial z-index is used.
     *
     * @param uuid - unique identifier of the overlay.
     * @returns
     */
    OverlayController.prototype.getZIndex = function (uuid) {
        return this.overlays.get(uuid) || this.initialZIndex;
    };
    /**
     * Removes given overlay from the overlays map.
     *
     * @param uuid - unique identifier of the overlay
     */
    OverlayController.prototype.removeOverlay = function (uuid) {
        this.overlays.delete(uuid);
    };
    return OverlayController;
}());
export { OverlayController };
