// @gooddata/sdk-ui-kit/esm/Dialog/ShareDialog/ShareDialogBase/GranularPermissions/permissionsLogic.js
// (C) 2023 GoodData Corporation
import { __read, __spread } from "tslib";
import invariant from "ts-invariant";
import { isGranularGranteeUser, } from "../types";
import { granularPermissionMessageTooltips } from "../../../../locales";
var allPermissionsSorted = ["EDIT", "SHARE", "VIEW"];
export var getEffectivePermission = function (permissions, inheritedPermissions, isSharedObjectLocked) {
    var allUserPermissions = __spread(permissions, inheritedPermissions);
    var effectivePermission = allPermissionsSorted.find(function (permission) {
        return allUserPermissions.includes(permission);
    });
    invariant(effectivePermission, "Provided item permissions are incomplete or invalid");
    if (isSharedObjectLocked && effectivePermission === "EDIT") {
        return "SHARE";
    }
    return effectivePermission;
};
var disableWithTooltip = function (possibility, tooltip) {
    possibility.enabled = false;
    possibility.tooltip = tooltip;
};
var getPermissionTypeItems = function (grantee, currentUserPermissions, isSharedObjectLocked) {
    return allPermissionsSorted.map(function (permission, permissionIndex) {
        var item = {
            id: permission,
            enabled: true,
            tooltip: "",
            hidden: false,
        };
        // don't allow setting permissions higher that the current user
        if ((permission === "EDIT" && !currentUserPermissions.canEditAffectedObject) ||
            (permission === "SHARE" && !currentUserPermissions.canShareAffectedObject) ||
            !currentUserPermissions.canViewAffectedObject) {
            disableWithTooltip(item, "shareDialog.share.granular.grantee.tooltip.cannotGrantHigher");
        }
        // hide Edit & share on inherited object
        if (permission === "EDIT" && isSharedObjectLocked) {
            item.hidden = true;
        }
        // don't allow setting permission lower that permission obtained indirectly
        grantee.inheritedPermissions.forEach(function (inheritedPermission) {
            var inheritedPermissionIndex = allPermissionsSorted.indexOf(inheritedPermission);
            if (permissionIndex > inheritedPermissionIndex) {
                disableWithTooltip(item, isGranularGranteeUser(grantee)
                    ? granularPermissionMessageTooltips.cannotGrantLowerForUser.id
                    : granularPermissionMessageTooltips.cannotGrantLowerForGroup.id);
            }
        });
        return item;
    });
};
export var getGranteePossibilities = function (grantee, currentUserPermissions, isSharedObjectLocked) {
    var granteeEffectivePermission = getEffectivePermission(grantee.permissions, grantee.inheritedPermissions, isSharedObjectLocked);
    // the "Remove" option state
    var remove = {
        enabled: true,
        tooltip: "",
    };
    // state of the whole permissions selection dropdown
    var change = {
        enabled: true,
        tooltip: "",
    };
    //state of the permissions selection dropdown items
    var permissionTypeItems = getPermissionTypeItems(grantee, currentUserPermissions, isSharedObjectLocked);
    // cannot change or remove permissions of a grantee that has higher permission than the current user
    if ((granteeEffectivePermission === "EDIT" && !currentUserPermissions.canEditAffectedObject) ||
        (granteeEffectivePermission === "SHARE" && !currentUserPermissions.canShareAffectedObject)) {
        var tooltipId = isGranularGranteeUser(grantee)
            ? granularPermissionMessageTooltips.cannotChangeHigherForUser.id
            : granularPermissionMessageTooltips.cannotChangeHigherForGroup.id;
        disableWithTooltip(change, tooltipId);
        disableWithTooltip(remove, tooltipId);
    }
    // cannot remove permission that is defined on the object in parent workspace
    if (grantee.permissions.length === 0 && grantee.inheritedPermissions.length !== 0) {
        disableWithTooltip(remove, isGranularGranteeUser(grantee)
            ? granularPermissionMessageTooltips.cannotRemoveFromParentForUser.id
            : granularPermissionMessageTooltips.cannotRemoveFromParentForGroup.id);
    }
    // disable all permissions change if all assignment options are disabled and also the Remove option is disabled
    if (change.enabled &&
        !remove.enabled &&
        permissionTypeItems.every(function (item) { return !item.enabled || item.hidden || item.id === granteeEffectivePermission; })) {
        disableWithTooltip(change, isGranularGranteeUser(grantee)
            ? granularPermissionMessageTooltips.noChangeAvailableForUser.id
            : granularPermissionMessageTooltips.noChangeAvailableForGroup.id);
    }
    return {
        remove: remove,
        assign: {
            items: permissionTypeItems,
            effective: granteeEffectivePermission,
        },
        change: change,
    };
};
