// @gooddata/sdk-ui-filters/esm/DateFilter/AbsoluteDateFilterForm/conversions.js
// (C) 2007-2022 GoodData Corporation
import { convertDateToPlatformDateString, convertPlatformDateStringToDate } from "../utils/DateConversions";
import { platformDateFormat, platformDateTimeFormat } from "../constants/Platform";
export var dateRangeToDateFilterValue = function (range, localIdentifier, isTimeForAbsoluteRangeEnabled) {
    var parsingFormat = isTimeForAbsoluteRangeEnabled ? platformDateTimeFormat : platformDateFormat;
    return {
        from: convertDateToPlatformDateString(range.from, parsingFormat),
        to: convertDateToPlatformDateString(range.to, parsingFormat),
        localIdentifier: localIdentifier,
        type: "absoluteForm",
        name: "",
        visible: true,
    };
};
export var dateFilterValueToDateRange = function (value, isTimeForAbsoluteRangeEnabled) {
    if (isTimeForAbsoluteRangeEnabled === void 0) { isTimeForAbsoluteRangeEnabled = false; }
    var isFromTimeDefined = value.from && value.from.split(" ").length > 1;
    var isToTimeDefined = value.to && value.to.split(" ").length > 1;
    var from = value && convertPlatformDateStringToDate(value.from);
    var to = value && convertPlatformDateStringToDate(value.to);
    if (from && !isFromTimeDefined && isTimeForAbsoluteRangeEnabled) {
        from.setHours(0);
        from.setMinutes(0);
    }
    if (to && !isToTimeDefined && isTimeForAbsoluteRangeEnabled) {
        to.setHours(23);
        to.setMinutes(59);
    }
    return { from: from, to: to };
};
