// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/chartCodeGenUtils.js
import { __assign } from "tslib";
// (C) 2022 GoodData Corporation
import { factoryNotationFor, isColorMappingItem, insightProperties, } from "@gooddata/sdk-model";
import filter from "lodash/fp/filter";
import flow from "lodash/fp/flow";
import fromPairs from "lodash/fromPairs";
import toPairs from "lodash/toPairs";
import { insightConversion, } from "../../utils/embeddingCodeGenerator";
import { getChartSupportedControls } from "../../utils/propertiesHelper";
import { removeUseless } from "../../utils/removeUseless";
var supportedChartConfigProperties = new Set([
    "colorMapping",
    "colorPalette",
    "dataLabels",
    "dataPoints",
    "dualAxis",
    "enableChartSorting",
    "enableSeparateTotalLabels",
    "enableJoinedAttributeAxisName",
    "grid",
    "legend",
    "legendLayout",
    "limits",
    "primaryChartType",
    "secondary_xaxis",
    "secondary_yaxis",
    "secondaryChartType",
    "separators",
    "stackMeasures",
    "stackMeasuresToPercent",
    "xaxis",
    "xFormat",
    "xLabel",
    "yaxis",
    "yFormat",
    "yLabel",
]);
export function chartConfigFromInsight(insight, ctx) {
    var _a, _b, _c, _d, _e, _f, _g;
    var properties = insightProperties(insight);
    var controls = (_a = properties === null || properties === void 0 ? void 0 : properties.controls) !== null && _a !== void 0 ? _a : {};
    var includeColorPalette = (_c = (_b = ctx === null || ctx === void 0 ? void 0 : ctx.backend) === null || _b === void 0 ? void 0 : _b.capabilities.supportsCustomColorPalettes) !== null && _c !== void 0 ? _c : false;
    var withValuesFromContext = __assign(__assign(__assign(__assign(__assign(__assign({}, controls), ((ctx === null || ctx === void 0 ? void 0 : ctx.colorPalette) && includeColorPalette ? { colorPalette: ctx.colorPalette } : {})), (((_d = ctx === null || ctx === void 0 ? void 0 : ctx.settings) === null || _d === void 0 ? void 0 : _d.separators) ? { separators: ctx.settings.separators } : {})), (((_e = ctx === null || ctx === void 0 ? void 0 : ctx.settings) === null || _e === void 0 ? void 0 : _e.enableChartsSorting) ? { enableChartSorting: true } : {})), (((_f = ctx === null || ctx === void 0 ? void 0 : ctx.settings) === null || _f === void 0 ? void 0 : _f.enableSeparateTotalLabels) ? { enableSeparateTotalLabels: true } : {})), (((_g = ctx === null || ctx === void 0 ? void 0 : ctx.settings) === null || _g === void 0 ? void 0 : _g.enableAxisNameViewByTwoAttributes) ? { enableJoinedAttributeAxisName: true } : {}));
    return flow(toPairs, filter(function (_a) {
        var key = _a[0];
        return supportedChartConfigProperties.has(key);
    }), fromPairs, function (c) { return getChartSupportedControls(c, insight, ctx === null || ctx === void 0 ? void 0 : ctx.settings); }, removeUseless)(withValuesFromContext);
}
export function chartAdditionalFactories(options) {
    var _a = (options !== null && options !== void 0 ? options : {}).getColorMappingPredicatePackage, getColorMappingPredicatePackage = _a === void 0 ? "@gooddata/sdk-ui-charts" : _a;
    return [
        {
            importInfo: {
                importType: "named",
                name: "getColorMappingPredicate",
                package: getColorMappingPredicatePackage,
            },
            transformation: function (obj) {
                return isColorMappingItem(obj)
                    ? "{predicate: getColorMappingPredicate(\"" + obj.id + "\"), color: " + factoryNotationFor(obj.color) + "}"
                    : undefined;
            },
        },
    ];
}
var chartConfigPropMeta = {
    typeImport: {
        importType: "named",
        name: "IChartConfig",
        package: "@gooddata/sdk-ui-charts",
    },
    cardinality: "scalar",
};
export function chartConfigInsightConversion(propName) {
    return insightConversion(propName, chartConfigPropMeta, chartConfigFromInsight);
}
