// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/comboChart/ComboChartDescriptor.js
import { __extends } from "tslib";
import { BucketNames } from "@gooddata/sdk-ui";
import { PluggableComboChart } from "./PluggableComboChart";
import { BigChartDescriptor } from "../BigChartDescriptor";
import { getReactEmbeddingCodeGenerator, getInsightToPropsConverter, filtersInsightConversion, sortsInsightConversion, multipleMeasuresBucketConversion, multipleAttributesBucketConversion, localeInsightConversion, executionConfigInsightConversion, } from "../../../utils/embeddingCodeGenerator";
import { chartAdditionalFactories, chartConfigInsightConversion } from "../chartCodeGenUtils";
var ComboChartDescriptor = /** @class */ (function (_super) {
    __extends(ComboChartDescriptor, _super);
    function ComboChartDescriptor() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.getEmbeddingCode = getReactEmbeddingCodeGenerator({
            component: {
                importType: "named",
                name: "ComboChart",
                package: "@gooddata/sdk-ui-charts",
            },
            insightToProps: getInsightToPropsConverter({
                primaryMeasures: multipleMeasuresBucketConversion("primaryMeasures", BucketNames.MEASURES),
                secondaryMeasures: multipleMeasuresBucketConversion("secondaryMeasures", BucketNames.SECONDARY_MEASURES),
                viewBy: multipleAttributesBucketConversion("viewBy", BucketNames.VIEW),
                filters: filtersInsightConversion("filters"),
                sortBy: sortsInsightConversion("sortBy"),
                config: chartConfigInsightConversion("config"),
                locale: localeInsightConversion("locale"),
                execConfig: executionConfigInsightConversion("execConfig"),
            }),
            additionalFactories: chartAdditionalFactories(),
        });
        return _this;
    }
    ComboChartDescriptor.prototype.getFactory = function () {
        return function (params) { return new PluggableComboChart(params); };
    };
    ComboChartDescriptor.prototype.getMeta = function () {
        return {
            documentationUrl: "https://sdk.gooddata.com/gooddata-ui/docs/combo_chart_component.html",
            supportsExport: true,
        };
    };
    return ComboChartDescriptor;
}(BigChartDescriptor));
export { ComboChartDescriptor };
