// @gooddata/sdk-ui-filters/esm/AttributeFilterHandler/internal/redux/loadAttribute/loadAttributeReducers.js
import identity from "lodash/identity";
var loadAttributeRequest = identity;
var loadAttributeStart = function (state) {
    state.attribute.status = "loading";
    state.attribute.error = undefined;
    state.attribute.data = undefined;
};
var loadAttributeSuccess = function (state, action) {
    state.attribute.status = "success";
    state.attribute.data = action.payload.attribute;
};
var loadAttributeError = function (state, action) {
    state.attribute.status = "error";
    state.attribute.error = action.payload.error;
};
var loadAttributeCancelRequest = identity;
var loadAttributeCancel = function (state) {
    state.attribute.status = "canceled";
};
/**
 * @internal
 */
export var loadAttributeReducers = {
    loadAttributeRequest: loadAttributeRequest,
    loadAttributeStart: loadAttributeStart,
    loadAttributeSuccess: loadAttributeSuccess,
    loadAttributeError: loadAttributeError,
    loadAttributeCancelRequest: loadAttributeCancelRequest,
    loadAttributeCancel: loadAttributeCancel,
};
