// @gooddata/sdk-ui-kit/esm/Dialog/ShareDialog/ShareDialogBase/types.js
import isEmpty from "lodash/isEmpty";
/**
 * @internal
 */
export var isGranteeUser = function (obj) {
    return !isEmpty(obj) && obj.type === "user";
};
/**
 * @internal
 */
export var isGranularGranteeUser = function (obj) {
    return !isEmpty(obj) && obj.type === "granularUser";
};
/**
 * @internal
 */
export var isGranteeUserInactive = function (obj) {
    return !isEmpty(obj) && obj.type === "inactive_owner";
};
/**
 * @internal
 */
export var isGranteeGroup = function (obj) {
    return !isEmpty(obj) && obj.type === "group";
};
/**
 * @internal
 */
export var isGranularGranteeGroup = function (obj) {
    return !isEmpty(obj) && obj.type === "granularGroup";
};
/**
 * @internal
 */
export var isGranularGrantee = function (obj) {
    return isGranularGranteeUser(obj) || isGranularGranteeGroup(obj);
};
/**
 * @internal
 */
export var isGranteeGroupAll = function (obj) {
    return !isEmpty(obj) && obj.type === "groupAll";
};
/**
 * @internal
 */
export var isGranteeItem = function (obj) {
    return (!isEmpty(obj) &&
        (isGranteeGroupAll(obj) || isGranteeGroup(obj) || isGranteeUserInactive(obj) || isGranteeUser(obj)));
};
/**
 * @internal
 */
export var isSelectErrorOption = function (obj) {
    return !isEmpty(obj) && obj.type === "error";
};
