// @gooddata/sdk-ui-kit/esm/Messages/useToastMessage.js
// (C) 2021-2022 GoodData Corporation
import { __assign } from "tslib";
import { useContext } from "react";
import { useIntl } from "react-intl";
import { ToastMessageContext } from "./ToastMessageContext";
/**
 * @internal
 */
export var useToastMessage = function () {
    var _a = useContext(ToastMessageContext), addMessage = _a.addMessage, removeMessage = _a.removeMessage, removeAllMessages = _a.removeAllMessages;
    var intl = useIntl();
    var addMessageBase = function (type) {
        return function (message, options) {
            if (options === null || options === void 0 ? void 0 : options.values) {
                return addMessage(__assign(__assign({}, options), { type: type, node: intl.formatMessage(message, options.values) }));
            }
            return addMessage(__assign(__assign({}, options), { type: type, text: intl.formatMessage(message) }));
        };
    };
    var addSuccess = addMessageBase("success");
    var addProgress = addMessageBase("progress");
    var addWarning = addMessageBase("warning");
    var addError = addMessageBase("error");
    return {
        addSuccess: addSuccess,
        addProgress: addProgress,
        addWarning: addWarning,
        addError: addError,
        removeMessage: removeMessage,
        removeAllMessages: removeAllMessages,
    };
};
