// @gooddata/sdk-ui-dashboard/esm/presentation/widget/widget/useInsightMenu.js
// (C) 2021-2022 GoodData Corporation
import { __read } from "tslib";
import { useCallback, useMemo, useState } from "react";
import { useIntl } from "react-intl";
import { selectExecutionResultByRef, useDashboardSelector } from "../../../model";
import { isDataError } from "../../../_staging/errors/errorPredicates";
import { useDashboardCustomizationsContext } from "../../dashboardContexts";
import { getDefaultInsightMenuItems, getDefaultLegacyInsightMenuItems, } from "../insightMenu";
export var useInsightMenu = function (config) {
    var exportCSVEnabled = config.exportCSVEnabled, exportXLSXEnabled = config.exportXLSXEnabled, scheduleExportEnabled = config.scheduleExportEnabled, insight = config.insight, onExportCSV = config.onExportCSV, onExportXLSX = config.onExportXLSX, onScheduleExport = config.onScheduleExport, isScheduleExportVisible = config.isScheduleExportVisible, widget = config.widget;
    var intl = useIntl();
    var _a = __read(useState(false), 2), isMenuOpen = _a[0], setIsMenuOpen = _a[1];
    var closeMenu = useCallback(function () { return setIsMenuOpen(false); }, []);
    var openMenu = useCallback(function () { return setIsMenuOpen(true); }, []);
    var execution = useDashboardSelector(selectExecutionResultByRef(widget.ref));
    var insightMenuItemsProvider = useDashboardCustomizationsContext().insightMenuItemsProvider;
    var defaultMenuItems = useMemo(function () {
        var useLegacyMenu = !insightMenuItemsProvider;
        var defaultMenuItemsGetter = useLegacyMenu
            ? getDefaultLegacyInsightMenuItems
            : getDefaultInsightMenuItems;
        return defaultMenuItemsGetter(intl, {
            exportCSVDisabled: !exportCSVEnabled,
            exportXLSXDisabled: !exportXLSXEnabled,
            scheduleExportDisabled: !scheduleExportEnabled,
            onExportCSV: function () {
                setIsMenuOpen(false);
                onExportCSV();
            },
            onExportXLSX: function () {
                setIsMenuOpen(false);
                onExportXLSX();
            },
            onScheduleExport: function () {
                setIsMenuOpen(false);
                onScheduleExport();
            },
            isScheduleExportVisible: isScheduleExportVisible,
            tooltipMessage: isDataError(execution === null || execution === void 0 ? void 0 : execution.error)
                ? intl.formatMessage({ id: "options.menu.unsupported.error" })
                : intl.formatMessage({ id: "options.menu.unsupported.loading" }),
        });
    }, [
        insightMenuItemsProvider,
        execution,
        exportCSVEnabled,
        exportXLSXEnabled,
        scheduleExportEnabled,
        onExportCSV,
        onExportXLSX,
        onScheduleExport,
        isScheduleExportVisible,
        intl,
    ]);
    var menuItems = useMemo(function () {
        return insightMenuItemsProvider
            ? insightMenuItemsProvider(insight, widget, defaultMenuItems, closeMenu)
            : defaultMenuItems;
    }, [insightMenuItemsProvider, insight, widget, defaultMenuItems, closeMenu]);
    return { menuItems: menuItems, isMenuOpen: isMenuOpen, openMenu: openMenu, closeMenu: closeMenu };
};
