// @gooddata/sdk-backend-spi/esm/deprecated/dateFilterConfigs/index.js
// (C) 2007-2022 GoodData Corporation
/* eslint-disable @typescript-eslint/no-empty-interface, import/namespace */
import * as m from "@gooddata/sdk-model";
/**
 * Type-guard testing whether the provided object is an instance of {@link DateFilterGranularity}.
 * @deprecated Use {@link @gooddata/sdk-model#isDateFilterGranularity}
 * @alpha
 */
export var isDateFilterGranularity = m.isDateFilterGranularity;
/**
 * Type-guard testing whether the provided object is an instance of {@link IAllTimeDateFilterOption}.
 * @deprecated Use {@link @gooddata/sdk-model#isAllTimeDateFilterOption}
 * @alpha
 */
export var isAllTimeDateFilterOption = m.isAllTimeDateFilterOption;
/**
 * Type-guard testing whether the provided object is an instance of {@link IAbsoluteDateFilterForm}.
 * @deprecated Use {@link @gooddata/sdk-model#isAbsoluteDateFilterForm}
 * @alpha
 */
export var isAbsoluteDateFilterForm = m.isAbsoluteDateFilterForm;
/**
 * Type-guard testing whether the provided object is an instance of {@link IAbsoluteDateFilterPreset}.
 * @deprecated Use {@link @gooddata/sdk-model#isAbsoluteDateFilterPreset}
 * @alpha
 */
export var isAbsoluteDateFilterPreset = m.isAbsoluteDateFilterPreset;
/**
 * Type-guard testing whether the provided object is an instance of {@link IRelativeDateFilterForm}.
 * @deprecated Use {@link @gooddata/sdk-model#isRelativeDateFilterForm}
 * @alpha
 */
export var isRelativeDateFilterForm = m.isRelativeDateFilterForm;
/**
 * Type-guard testing whether the provided object is an instance of {@link IRelativeDateFilterPreset}.
 * @deprecated Use {@link @gooddata/sdk-model#isRelativeDateFilterPreset}
 * @alpha
 */
export var isRelativeDateFilterPreset = m.isRelativeDateFilterPreset;
