// @gooddata/sdk-ui-kit/esm/Dialog/EmbedInsightDialog/utils.js
// (C) 2022 GoodData Corporation
import { DEFAULT_HEIGHT } from "./EmbedInsightDialogBase/types";
/**
 * @internal
 */
export var getDefaultEmbedCodeOptions = function (codeType) {
    if (codeType === "definition") {
        return {
            type: "definition",
            includeConfiguration: true,
            customHeight: true,
        };
    }
    return {
        type: "reference",
        displayTitle: true,
        customHeight: true,
    };
};
/**
 * @internal
 */
export var getHeightWithUnitsForEmbedCode = function (codeOption) {
    var unit = codeOption.unit ? codeOption.unit : "px";
    var customHeightValue = codeOption.height
        ? codeOption.height
        : getDefaultHeightForEmbedCodeByUnit(unit);
    var height = codeOption.customHeight ? customHeightValue : getDefaultHeightForEmbedCodeByUnit(unit);
    if (unit === "px") {
        return Number(height);
    }
    return "" + height + unit;
};
/**
 * @internal
 */
export var getDefaultHeightForEmbedCodeByUnit = function (unit) {
    return DEFAULT_HEIGHT[unit];
};
