// @gooddata/sdk-backend-spi/esm/deprecated/dashboard/drills.js
// (C) 2020-2022 GoodData Corporation
/* eslint-disable @typescript-eslint/no-empty-interface, import/namespace */
import * as m from "@gooddata/sdk-model";
export { getAttributeIdentifiersPlaceholdersFromUrl } from "./drillUrls";
/**
 * Type-guard testing whether the provided object is an instance of {@link IDrillFromMeasure}.
 * @deprecated Use {@link @gooddata/sdk-model#isDrillFromMeasure}
 * @alpha
 */
export var isDrillFromMeasure = m.isDrillFromMeasure;
/**
 * Type-guard testing whether the provided object is an instance of {@link IDrillFromAttribute}.
 * @deprecated Use {@link @gooddata/sdk-model#isDrillFromAttribute}
 * @alpha
 */
export var isDrillFromAttribute = m.isDrillFromAttribute;
/**
 * Type-guard testing whether the provided object is an instance of {@link IDrillToLegacyDashboard}.
 * @deprecated Use {@link @gooddata/sdk-model#isDrillToLegacyDashboard}
 * @alpha
 */
export var isDrillToLegacyDashboard = m.isDrillToLegacyDashboard;
/**
 * Type-guard testing whether the provided object is an instance of {@link IDrillToDashboard}.
 * @deprecated Use {@link @gooddata/sdk-model#isDrillToDashboard}
 * @alpha
 */
export var isDrillToDashboard = m.isDrillToDashboard;
/**
 * Type-guard testing whether the provided object is an instance of {@link IDrillToInsight}.
 * @deprecated Use {@link @gooddata/sdk-model#isDrillToInsight}
 * @alpha
 */
export var isDrillToInsight = m.isDrillToInsight;
/**
 * Type-guard testing whether the provided object is an instance of {@link IDrillToCustomUrl}.
 * @deprecated Use {@link @gooddata/sdk-model#isDrillToCustomUrl}
 * @alpha
 */
export var isDrillToCustomUrl = m.isDrillToCustomUrl;
/**
 * Type-guard testing whether the provided object is an instance of {@link IDrillToAttributeUrl}.
 * @deprecated Use {@link @gooddata/sdk-model#isDrillToAttributeUrl}
 * @alpha
 */
export var isDrillToAttributeUrl = m.isDrillToAttributeUrl;
