// @gooddata/api-model-bear/esm/visualizationObject/utils.js
import { __spreadArrays } from "tslib";
// (C) 2007-2022 GoodData Corporation
import { GdcVisualizationObject } from "./GdcVisualizationObject";
var isAttribute = GdcVisualizationObject.isAttribute;
var isMeasure = GdcVisualizationObject.isMeasure;
var isAttributeFilter = GdcVisualizationObject.isAttributeFilter;
function getAttributesInBucket(bucket) {
    return bucket.items.reduce(function (list, bucketItem) {
        if (isAttribute(bucketItem)) {
            list.push(bucketItem.visualizationAttribute);
        }
        return list;
    }, []);
}
function getAttributes(mdObject) {
    var buckets = mdObject.buckets;
    return buckets.reduce(function (categoriesList, bucket) {
        categoriesList.push.apply(categoriesList, getAttributesInBucket(bucket));
        return categoriesList;
    }, []);
}
function getMeasuresInBucket(bucket) {
    return bucket.items.reduce(function (list, bucketItem) {
        if (isMeasure(bucketItem)) {
            list.push(bucketItem.measure);
        }
        return list;
    }, []);
}
function getDefinition(measure) {
    return GdcVisualizationObject.isMeasureDefinition(measure.definition)
        ? measure.definition.measureDefinition
        : undefined;
}
function getMeasures(mdObject) {
    var buckets = mdObject.buckets;
    return buckets.reduce(function (measuresList, bucket) {
        measuresList.push.apply(measuresList, getMeasuresInBucket(bucket));
        return measuresList;
    }, []);
}
function getMeasureFilters(measure) {
    var _a, _b;
    return (_b = (_a = getDefinition(measure)) === null || _a === void 0 ? void 0 : _a.filters) !== null && _b !== void 0 ? _b : [];
}
function getMeasureAttributeFilters(measure) {
    return getMeasureFilters(measure).filter(isAttributeFilter);
}
function getAttributeFilters(mdObject) {
    return getMeasures(mdObject).reduce(function (filters, measure) {
        filters.push.apply(filters, getMeasureAttributeFilters(measure));
        return filters;
    }, []);
}
function getAttributeFilterDisplayForm(measureFilter) {
    return GdcVisualizationObject.isPositiveAttributeFilter(measureFilter)
        ? measureFilter.positiveAttributeFilter.displayForm.uri
        : measureFilter.negativeAttributeFilter.displayForm.uri;
}
/**
 * @public
 */
export function getAttributesDisplayForms(mdObject) {
    var attributesDfs = getAttributes(mdObject).map(function (attribute) {
        return attribute.displayForm.uri;
    });
    var attrMeasureFilters = getAttributeFilters(mdObject);
    var attrMeasureFiltersDfs = attrMeasureFilters.map(getAttributeFilterDisplayForm);
    return __spreadArrays(attrMeasureFiltersDfs, attributesDfs);
}
