// @gooddata/sdk-ui-dashboard/esm/presentation/dashboard/components/DateFilterConfigWarnings.js
import { __assign } from "tslib";
// (C) 2022 GoodData Corporation
import React from "react";
import { defineMessage, FormattedMessage } from "react-intl";
import { Message } from "@gooddata/sdk-ui-kit";
import { selectDateFilterConfigValidationWarnings, selectEnableRenamingProjectToWorkspace, selectIsInEditMode, useDashboardSelector, } from "../../../model";
var workspaceValidationMessagesMapping = {
    ConflictingIdentifiers: defineMessage({ id: "filters.config.warning.conflictingIdentifiers" }),
    NO_CONFIG: defineMessage({ id: "filters.config.warning.workspace.notAvailable" }),
    NoVisibleOptions: defineMessage({ id: "filters.config.warning.allOptionsHidden" }),
    SelectedOptionInvalid: defineMessage({ id: "filters.config.warning.selectedFilterNotValid" }),
    TOO_MANY_CONFIGS: defineMessage({ id: "filters.config.warning.multipleWorkspacesConfigs" }),
};
var projectValidationMessagesMapping = __assign(__assign({}, workspaceValidationMessagesMapping), { NO_CONFIG: defineMessage({ id: "filters.config.warning.notAvailable" }), TOO_MANY_CONFIGS: defineMessage({ id: "filters.config.warning.multipleProjectConfigs" }) });
export var DateFilterConfigWarnings = function () {
    var isInEditMode = useDashboardSelector(selectIsInEditMode);
    var warnings = useDashboardSelector(selectDateFilterConfigValidationWarnings);
    var enableRenamingProjectToWorkspace = useDashboardSelector(selectEnableRenamingProjectToWorkspace);
    var effectiveMapping = enableRenamingProjectToWorkspace
        ? workspaceValidationMessagesMapping
        : projectValidationMessagesMapping;
    return isInEditMode && warnings.length > 0 ? (React.createElement(Message, { type: "warning", contrast: true, className: "gd-date-filter-config-warning-message" },
        React.createElement("ul", { className: "gd-date-filter-config-warning-message-items" }, warnings.map(function (warning) {
            var message = effectiveMapping[warning];
            if (message) {
                return (React.createElement("li", { key: warning },
                    React.createElement(FormattedMessage, { id: message.id })));
            }
            return null;
        })))) : null;
};
